// @aztec/sdk/user_tx/user_join_split_tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserJoinSplitTx = void 0;
const client_proofs_1 = require("@aztec/barretenberg/client_proofs");
class UserJoinSplitTx {
    constructor(txHash, userId, assetId, publicInput, publicOutput, privateInput, recipientPrivateOutput, senderPrivateOutput, inputOwner, outputOwner, ownedByUser, created, settled) {
        this.txHash = txHash;
        this.userId = userId;
        this.assetId = assetId;
        this.publicInput = publicInput;
        this.publicOutput = publicOutput;
        this.privateInput = privateInput;
        this.recipientPrivateOutput = recipientPrivateOutput;
        this.senderPrivateOutput = senderPrivateOutput;
        this.inputOwner = inputOwner;
        this.outputOwner = outputOwner;
        this.ownedByUser = ownedByUser;
        this.created = created;
        this.settled = settled;
        this.proofId = client_proofs_1.ProofId.JOIN_SPLIT;
    }
}
exports.UserJoinSplitTx = UserJoinSplitTx;
