// @aztec/sdk/ethereum_sdk/database/dexie_database.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DexieDatabase = void 0;
const tslib_1 = require("tslib");
const address_1 = require("@aztec/barretenberg/address");
const dexie_1 = tslib_1.__importDefault(require("dexie"));
class DexieAccount {
    constructor(ethAddress, accountPublicKey) {
        this.ethAddress = ethAddress;
        this.accountPublicKey = accountPublicKey;
    }
}
const dexieAccountToDbAccount = ({ ethAddress, accountPublicKey }) => ({
    ethAddress: new address_1.EthAddress(Buffer.from(ethAddress)),
    accountPublicKey: new address_1.GrumpkinAddress(Buffer.from(accountPublicKey)),
});
class DexieDatabase {
    constructor(dbName = 'aztec2-sdk-eth', version = 1) {
        this.dbName = dbName;
        this.version = version;
    }
    async init() {
        this.createTables();
        // If schemas changed, delete db.
        try {
            await this.getAccount(address_1.EthAddress.ZERO);
        }
        catch (e) {
            await this.db.delete();
            this.createTables();
        }
    }
    createTables() {
        this.db = new dexie_1.default(this.dbName);
        this.db.version(this.version).stores({
            user: '&ethAddress',
        });
        this.user = this.db.table('user');
        this.user.mapToClass(DexieAccount);
    }
    async clear() {
        for (const table of this.db.tables) {
            await table.clear();
        }
    }
    async close() {
        await this.db.close();
    }
    async setAccount({ ethAddress, accountPublicKey }) {
        await this.user.put(new DexieAccount(new Uint8Array(ethAddress.toBuffer()), new Uint8Array(accountPublicKey.toBuffer())));
    }
    async getAccount(ethAddress) {
        const account = await this.user.get({ ethAddress: new Uint8Array(ethAddress.toBuffer()) });
        return account ? dexieAccountToDbAccount(account) : undefined;
    }
    async deleteAccount(ethAddress) {
        await this.user.delete(new Uint8Array(ethAddress.toBuffer()));
    }
}
exports.DexieDatabase = DexieDatabase;
