// @aztec/sdk/note_picker/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotePicker = void 0;
const sorted_notes_1 = require("./sorted_notes");
const pick_1 = require("./pick");
const noteSum = (notes) => notes.reduce((sum, { value }) => sum + value, BigInt(0));
class NotePicker {
    constructor(notes = [], excludeIndices = []) {
        this.notes = notes;
        this.numNotesPerTx = 2;
        const availableNotes = notes.filter(({ index }) => excludeIndices.indexOf(index) < 0);
        this.spendableNotes = new sorted_notes_1.SortedNotes(availableNotes);
    }
    getSpendableNotes(excludeNullifiers) {
        return (excludeNullifiers === null || excludeNullifiers === void 0 ? void 0 : excludeNullifiers.length) ? this.spendableNotes.filter(({ nullifier }) => !excludeNullifiers.some(n => n.equals(nullifier)))
            : this.spendableNotes;
    }
    pick(value, excludeNullifiers) {
        const spendableNotes = this.getSpendableNotes(excludeNullifiers);
        return pick_1.pick(spendableNotes, value);
    }
    getSum() {
        return noteSum(this.notes);
    }
    getSpendableSum(excludeNullifiers) {
        const spendableNotes = this.getSpendableNotes(excludeNullifiers);
        return noteSum(spendableNotes.notes);
    }
    getMaxSpendableValue(excludeNullifiers) {
        const spendableNotes = this.getSpendableNotes(excludeNullifiers);
        const notes = spendableNotes.slice(-this.numNotesPerTx);
        return noteSum(notes);
    }
}
exports.NotePicker = NotePicker;
