// @aztec/sdk/note_picker/sorted_notes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SortedNotes = void 0;
class SortedNotes {
    constructor(notes = []) {
        this.sortedNotes = [];
        this.bulkAdd(notes);
    }
    get length() {
        return this.sortedNotes.length;
    }
    get notes() {
        return [...this.sortedNotes];
    }
    add(note) {
        let i = this.sortedNotes.length;
        for (; i > 0; i--) {
            if (this.nth(i - 1).value <= note.value) {
                break;
            }
        }
        this.sortedNotes.splice(i, 0, note);
    }
    bulkAdd(notes) {
        if (!this.sortedNotes.length) {
            this.sortedNotes = [...notes].sort((a, b) => (a.value < b.value ? -1 : a.value > b.value ? 1 : 0));
        }
        else {
            notes.forEach(n => this.add(n));
        }
    }
    remove(note) {
        const idx = this.sortedNotes.findIndex(n => n.index === note.index);
        this.sortedNotes.splice(idx, 1);
    }
    forEach(callback) {
        this.sortedNotes.forEach((note, i) => callback(note, i));
    }
    nth(idx) {
        return this.sortedNotes[idx];
    }
    slice(start, end) {
        return this.sortedNotes.slice(start, end);
    }
    filter(cb) {
        const chosenNotes = this.sortedNotes.filter(cb);
        return new SortedNotes(chosenNotes);
    }
}
exports.SortedNotes = SortedNotes;
