// @aztec/sdk/txs_state/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxsState = void 0;
const tslib_1 = require("tslib");
const debug_1 = tslib_1.__importDefault(require("debug"));
const events_1 = require("events");
const debug = debug_1.default('bb:txs_state');
class TxsState extends events_1.EventEmitter {
    constructor(explorer, latest = 5) {
        super();
        this.explorer = explorer;
        this.latest = latest;
        this.running = false;
        this.rollups = [];
        this.txs = [];
    }
    async fetchLatestRollups() {
        const rollups = await this.explorer.getLatestRollups(this.latest);
        const hasChanged = rollups.some(rollup => {
            const prev = this.rollups.find(r => r.id === rollup.id);
            return !prev || prev.status !== rollup.status;
        });
        if (hasChanged) {
            this.rollups = rollups;
            this.emit('rollups', this.rollups);
        }
    }
    async fetchLatestTxs() {
        const txs = await this.explorer.getLatestTxs(this.latest);
        const hasChanged = txs.some(tx => {
            var _a, _b;
            const prevTx = this.txs.find(t => t.txHash.equals(tx.txHash));
            return !prevTx || ((_a = prevTx.rollup) === null || _a === void 0 ? void 0 : _a.status) !== ((_b = tx.rollup) === null || _b === void 0 ? void 0 : _b.status);
        });
        if (hasChanged) {
            this.txs = txs;
            this.emit('txs', this.txs);
        }
    }
    async start() {
        this.running = true;
        while (this.running) {
            try {
                await Promise.all([this.fetchLatestRollups(), this.fetchLatestTxs()]);
            }
            catch (err) {
                debug(err);
            }
            await new Promise(resolve => setTimeout(resolve, 1000));
        }
    }
    stop() {
        this.running = false;
    }
    async getLatestRollups(count) {
        return this.explorer.getLatestRollups(count);
    }
    async getLatestTxs(count) {
        return this.explorer.getLatestTxs(count);
    }
    async getRollup(id) {
        return this.explorer.getRollup(id);
    }
    async getTx(txHash) {
        return this.explorer.getTx(txHash);
    }
}
exports.TxsState = TxsState;
