// @aztec/blockchain/asset/eth_asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthAsset = void 0;
const address_1 = require("@aztec/barretenberg/address");
const units_1 = require("../units");
/* eslint-disable @typescript-eslint/no-unused-vars */
class EthAsset {
    constructor(provider) {
        this.provider = provider;
        this.precision = 6;
    }
    getStaticInfo() {
        return {
            address: address_1.EthAddress.ZERO,
            name: 'Eth',
            symbol: 'ETH',
            decimals: 18,
            permitSupport: false,
            gasConstants: [5000, 0, 5000, 30000, 0, 30000, 30000],
        };
    }
    async getUserNonce(account) {
        return BigInt(0);
    }
    async balanceOf(account) {
        return BigInt(await this.provider.getBalance(account.toString()));
    }
    async allowance(owner, receiver) {
        throw new Error('Allowance unsupported for ETH.');
    }
    async approve(value, owner, receiver) {
        throw new Error('Approve unsupported for ETH.');
    }
    async mint(value, account) {
        throw new Error('Mint unsupported for ETH.');
    }
    fromBaseUnits(value, precision = this.precision) {
        return units_1.fromBaseUnits(value, 18, precision);
    }
    toBaseUnits(value) {
        return units_1.toBaseUnits(value, 18);
    }
}
exports.EthAsset = EthAsset;
