// @aztec/blockchain/asset/token_asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenAsset = void 0;
const providers_1 = require("@ethersproject/providers");
const tx_hash_1 = require("@aztec/barretenberg/tx_hash");
const ethers_1 = require("ethers");
const units_1 = require("../units");
const abi = [
    'function decimals() public view returns (uint8)',
    'function approve(address spender, uint256 amount) public returns (bool)',
    'function allowance(address owner, address spender) public view returns (uint256)',
    'function balanceOf(address account) public view returns (uint256)',
    'function mint(address _to, uint256 _value) public returns (bool)',
    'function name() public view returns (string)',
    'function symbol() public view returns (string)',
    'function nonces(address) public view returns(uint256)',
];
class TokenAsset {
    constructor(ethersProvider, info, minConfirmations = 1) {
        this.ethersProvider = ethersProvider;
        this.info = info;
        this.minConfirmations = minConfirmations;
        this.precision = 2;
        this.contract = new ethers_1.Contract(info.address.toString(), abi, ethersProvider);
    }
    static async fromAddress(address, ethersProvider, permitSupport, minConfirmations = 1) {
        const contract = new ethers_1.Contract(address.toString(), abi, ethersProvider);
        const info = {
            address,
            name: await contract.name(),
            symbol: await contract.symbol(),
            decimals: +(await contract.decimals()),
            permitSupport,
            gasConstants: [5000, 0, 36000, 36000, 0, 36000, 36000],
        };
        return new TokenAsset(ethersProvider, info, minConfirmations);
    }
    getStaticInfo() {
        return this.info;
    }
    async getUserNonce(account) {
        return BigInt(await this.contract.nonces(account.toString()));
    }
    async balanceOf(account) {
        const balance = await this.contract.balanceOf(account.toString());
        return BigInt(balance);
    }
    async allowance(owner, receiver) {
        const allowance = await this.contract.allowance(owner.toString(), receiver.toString());
        return BigInt(allowance);
    }
    async approve(value, owner, receiver, provider) {
        const contract = this.getContractWithSigner(owner, provider);
        const res = (await contract.approve(receiver.toString(), value));
        const receipt = await res.wait(this.minConfirmations);
        return tx_hash_1.TxHash.fromString(receipt.transactionHash);
    }
    async mint(value, account, provider) {
        const contract = this.getContractWithSigner(account, provider);
        const res = await contract.mint(account.toString(), value);
        const receipt = await res.wait(this.minConfirmations);
        return tx_hash_1.TxHash.fromString(receipt.transactionHash);
    }
    fromBaseUnits(value, precision = this.precision) {
        return units_1.fromBaseUnits(value, this.info.decimals, precision);
    }
    toBaseUnits(value) {
        return units_1.toBaseUnits(value, this.info.decimals);
    }
    getContractWithSigner(account, provider) {
        const ethSigner = (provider ? new providers_1.Web3Provider(provider) : this.ethersProvider).getSigner(account.toString());
        return new ethers_1.Contract(this.info.address.toString(), abi, ethSigner);
    }
}
exports.TokenAsset = TokenAsset;
