// @aztec/barretenberg/crypto/schnorr/signature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchnorrSignature = void 0;
const crypto_1 = require("crypto");
class SchnorrSignature {
    constructor(buffer) {
        this.buffer = buffer;
        if (buffer.length !== 64) {
            throw new Error('Invalid signature buffer.');
        }
    }
    static isSignature(signature) {
        return /^(0x)?[0-9a-f]{128}$/i.test(signature);
    }
    static fromString(signature) {
        if (!SchnorrSignature.isSignature(signature)) {
            throw new Error(`Invalid signature string: ${signature}`);
        }
        return new SchnorrSignature(Buffer.from(signature.replace(/^0x/i, ''), 'hex'));
    }
    static randomSignature() {
        return new SchnorrSignature(crypto_1.randomBytes(64));
    }
    s() {
        return this.buffer.slice(0, 32);
    }
    e() {
        return this.buffer.slice(32);
    }
    toBuffer() {
        return this.buffer;
    }
    toString() {
        return `0x${this.buffer.toString('hex')}`;
    }
}
exports.SchnorrSignature = SchnorrSignature;
