// @aztec/sdk/user/recovery_payload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecoveryPayload = exports.RecoveryData = void 0;
const address_1 = require("@aztec/barretenberg/address");
const account_id_1 = require("@aztec/barretenberg/account_id");
const crypto_1 = require("@aztec/barretenberg/crypto");
class RecoveryData {
    constructor(accountId, signature) {
        this.accountId = accountId;
        this.signature = signature;
    }
    static fromBuffer(data) {
        const accountId = account_id_1.AccountId.fromBuffer(data.slice(0, 32));
        const signature = new crypto_1.SchnorrSignature(data.slice(32, 96));
        return new RecoveryData(accountId, signature);
    }
    static fromString(data) {
        return RecoveryData.fromBuffer(Buffer.from(data.replace(/^0x/i, ''), 'hex'));
    }
    toBuffer() {
        return Buffer.concat([this.accountId.toBuffer(), this.signature.toBuffer()]);
    }
    toString() {
        return `0x${this.toBuffer().toString('hex')}`;
    }
}
exports.RecoveryData = RecoveryData;
class RecoveryPayload {
    constructor(trustedThirdPartyPublicKey, recoveryPublicKey, recoveryData) {
        this.trustedThirdPartyPublicKey = trustedThirdPartyPublicKey;
        this.recoveryPublicKey = recoveryPublicKey;
        this.recoveryData = recoveryData;
    }
    static fromBuffer(data) {
        const trustedThirdPartyPublicKey = new address_1.GrumpkinAddress(data.slice(0, 64));
        const recoveryPublicKey = new address_1.GrumpkinAddress(data.slice(64, 64 + 64));
        const recoveryData = RecoveryData.fromBuffer(data.slice(64 + 64));
        return new RecoveryPayload(trustedThirdPartyPublicKey, recoveryPublicKey, recoveryData);
    }
    static fromString(data) {
        return RecoveryPayload.fromBuffer(Buffer.from(data.replace(/^0x/i, ''), 'hex'));
    }
    toBuffer() {
        return Buffer.concat([
            this.trustedThirdPartyPublicKey.toBuffer(),
            this.recoveryPublicKey.toBuffer(),
            this.recoveryData.toBuffer(),
        ]);
    }
    toString() {
        return `0x${this.toBuffer().toString('hex')}`;
    }
}
exports.RecoveryPayload = RecoveryPayload;
