// @aztec/sdk/ethereum_sdk/database/sql_database/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLDatabase = exports.getOrmConfig = void 0;
const account_dao_1 = require("./account_dao");
exports.getOrmConfig = (dbPath) => ({
    name: 'aztec2-sdk-eth',
    type: 'sqlite',
    database: dbPath === ':memory:' ? dbPath : `${dbPath || '.'}/aztec2-sdk-eth.sqlite`,
    entities: [account_dao_1.AccountDao],
    synchronize: true,
    logging: false,
});
class SQLDatabase {
    constructor(connection) {
        this.connection = connection;
        this.accountRep = this.connection.getRepository(account_dao_1.AccountDao);
    }
    async init() { }
    async close() {
        await this.connection.close();
    }
    async clear() {
        await this.connection.synchronize(true);
    }
    async setAccount(account) {
        await this.accountRep.save(account);
    }
    async getAccount(ethAddress) {
        return this.accountRep.findOne({ ethAddress });
    }
    async deleteAccount(ethAddress) {
        await this.accountRep.delete({ ethAddress });
    }
}
exports.SQLDatabase = SQLDatabase;
