// @aztec/barretenberg/note_algorithms/claim_note_tx_data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaimNoteTxData = void 0;
const bigint_buffer_1 = require("../bigint_buffer");
const address_1 = require("../address");
const serialize_1 = require("../serialize");
const viewing_key_1 = require("../viewing_key");
const bridge_id_1 = require("../bridge_id");
const derive_note_secret_1 = require("./derive_note_secret");
class ClaimNoteTxData {
    constructor(value, bridgeId, ownerPubKey, ownerNonce, noteSecret) {
        this.value = value;
        this.bridgeId = bridgeId;
        this.ownerPubKey = ownerPubKey;
        this.ownerNonce = ownerNonce;
        this.noteSecret = noteSecret;
    }
    static createFromEphPriv(value, bridgeId, owner, ephPrivKey, grumpkin) {
        const noteSecret = derive_note_secret_1.deriveNoteSecret(owner.publicKey, ephPrivKey, grumpkin);
        return new ClaimNoteTxData(value, bridgeId, owner.publicKey, owner.nonce, noteSecret);
    }
    static createFromEphPub(value, bridgeId, owner, ephPubKey, ownerPrivKey, grumpkin) {
        const noteSecret = derive_note_secret_1.deriveNoteSecret(ephPubKey, ownerPrivKey, grumpkin);
        return new ClaimNoteTxData(value, bridgeId, owner.publicKey, owner.nonce, noteSecret);
    }
    toBuffer() {
        return Buffer.concat([
            bigint_buffer_1.toBufferBE(this.value, 32),
            this.bridgeId.toBuffer(),
            this.ownerPubKey.toBuffer(),
            serialize_1.numToUInt32BE(this.ownerNonce),
            this.noteSecret,
        ]);
    }
    equals(note) {
        return this.toBuffer().equals(note.toBuffer());
    }
    getViewingKey(ownerPubKey, ephPrivKey, grumpkin) {
        const noteBuf = Buffer.concat([bigint_buffer_1.toBufferBE(this.value, 32), this.bridgeId.toBuffer().slice(0, 8)]);
        return viewing_key_1.ViewingKey.createFromEphPriv(noteBuf, ownerPubKey, ephPrivKey, grumpkin);
    }
}
exports.ClaimNoteTxData = ClaimNoteTxData;
ClaimNoteTxData.EMPTY = new ClaimNoteTxData(BigInt(0), bridge_id_1.BridgeId.ZERO, address_1.GrumpkinAddress.one(), 0, Buffer.alloc(32));
