// @aztec/barretenberg/note_algorithms/defi_interaction_note.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.packInteractionNotes = exports.DefiInteractionNote = void 0;
const bigint_buffer_1 = require("../bigint_buffer");
const crypto_1 = require("crypto");
const serialize_1 = require("../serialize");
const bridge_id_1 = require("../bridge_id");
class DefiInteractionNote {
    constructor(bridgeId, nonce, totalInputValue, totalOutputValueA, totalOutputValueB, result) {
        this.bridgeId = bridgeId;
        this.nonce = nonce;
        this.totalInputValue = totalInputValue;
        this.totalOutputValueA = totalOutputValueA;
        this.totalOutputValueB = totalOutputValueB;
        this.result = result;
    }
    static random() {
        return new DefiInteractionNote(bridge_id_1.BridgeId.random(), crypto_1.randomBytes(4).readUInt32BE(), bigint_buffer_1.toBigIntBE(crypto_1.randomBytes(32)), bigint_buffer_1.toBigIntBE(crypto_1.randomBytes(32)), bigint_buffer_1.toBigIntBE(crypto_1.randomBytes(32)), !!Math.round(Math.random()));
    }
    static fromBuffer(buf) {
        const bridgeId = bridge_id_1.BridgeId.fromBuffer(buf.slice(0, 32));
        let offset = 32;
        const totalInputValue = bigint_buffer_1.toBigIntBE(buf.slice(offset, offset + 32));
        offset += 32;
        const totalOutputValueA = bigint_buffer_1.toBigIntBE(buf.slice(offset, offset + 32));
        offset += 32;
        const totalOutputValueB = bigint_buffer_1.toBigIntBE(buf.slice(offset, offset + 32));
        offset += 32;
        const nonce = buf.readUInt32BE(offset);
        offset += 4;
        const result = !!buf[offset];
        return new DefiInteractionNote(bridgeId, nonce, totalInputValue, totalOutputValueA, totalOutputValueB, result);
    }
    toBuffer() {
        return Buffer.concat([
            this.bridgeId.toBuffer(),
            bigint_buffer_1.toBufferBE(this.totalInputValue, 32),
            bigint_buffer_1.toBufferBE(this.totalOutputValueA, 32),
            bigint_buffer_1.toBufferBE(this.totalOutputValueB, 32),
            serialize_1.numToUInt32BE(this.nonce),
            Buffer.from([+this.result]),
        ]);
    }
    equals(note) {
        return this.toBuffer().equals(note.toBuffer());
    }
}
exports.DefiInteractionNote = DefiInteractionNote;
DefiInteractionNote.EMPTY = new DefiInteractionNote(bridge_id_1.BridgeId.ZERO, 0, BigInt(0), BigInt(0), BigInt(0), false);
DefiInteractionNote.LENGTH = DefiInteractionNote.EMPTY.toBuffer().length;
exports.packInteractionNotes = (notes) => {
    const hash = crypto_1.createHash('sha256')
        .update(Buffer.concat(notes.map(note => Buffer.concat([
        note.bridgeId.toBuffer(),
        serialize_1.numToUInt32BE(note.nonce, 32),
        bigint_buffer_1.toBufferBE(note.totalInputValue, 32),
        bigint_buffer_1.toBufferBE(note.totalOutputValueA, 32),
        bigint_buffer_1.toBufferBE(note.totalOutputValueB, 32),
        Buffer.alloc(31),
        Buffer.from([+note.result]),
    ]))))
        .digest();
    // Zero the first 4 bits.
    hash[0] &= 15;
    return hash;
};
