// @aztec/barretenberg/note_algorithms/tree_claim_note.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreeClaimNote = void 0;
const bigint_buffer_1 = require("../bigint_buffer");
const crypto_1 = require("crypto");
const bridge_id_1 = require("../bridge_id");
const serialize_1 = require("../serialize");
class TreeClaimNote {
    constructor(value, bridgeId, defiInteractionNonce, fee, partialState) {
        this.value = value;
        this.bridgeId = bridgeId;
        this.defiInteractionNonce = defiInteractionNonce;
        this.fee = fee;
        this.partialState = partialState;
    }
    static random() {
        return new TreeClaimNote(bigint_buffer_1.toBigIntBE(crypto_1.randomBytes(32)), bridge_id_1.BridgeId.random(), crypto_1.randomBytes(4).readUInt32BE(), bigint_buffer_1.toBigIntBE(crypto_1.randomBytes(32)), crypto_1.randomBytes(32));
    }
    static fromBuffer(buf) {
        const value = bigint_buffer_1.toBigIntBE(buf.slice(0, 32));
        let offset = 32;
        const bridgeId = bridge_id_1.BridgeId.fromBuffer(buf.slice(offset, offset + bridge_id_1.BridgeId.LENGTH));
        offset += 32;
        const defiInteractionNonce = buf.readUInt32BE(offset);
        offset += 4;
        const fee = bigint_buffer_1.toBigIntBE(buf.slice(offset, offset + 32));
        offset += 32;
        const partialState = buf.slice(offset, offset + 32);
        return new TreeClaimNote(value, bridgeId, defiInteractionNonce, fee, partialState);
    }
    toBuffer() {
        return Buffer.concat([
            bigint_buffer_1.toBufferBE(this.value, 32),
            this.bridgeId.toBuffer(),
            serialize_1.numToUInt32BE(this.defiInteractionNonce),
            bigint_buffer_1.toBufferBE(this.fee, 32),
            this.partialState,
        ]);
    }
    equals(note) {
        return this.toBuffer().equals(note.toBuffer());
    }
}
exports.TreeClaimNote = TreeClaimNote;
TreeClaimNote.EMPTY = new TreeClaimNote(BigInt(0), bridge_id_1.BridgeId.ZERO, 0, BigInt(0), Buffer.alloc(32));
TreeClaimNote.LENGTH = TreeClaimNote.EMPTY.toBuffer().length;
