// @aztec/barretenberg/note_algorithms/tree_note.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreeNote = void 0;
const bigint_buffer_1 = require("../bigint_buffer");
const address_1 = require("../address");
const serialize_1 = require("../serialize");
const viewing_key_1 = require("../viewing_key");
const derive_note_secret_1 = require("./derive_note_secret");
class TreeNote {
    constructor(ownerPubKey, value, assetId, nonce, noteSecret) {
        this.ownerPubKey = ownerPubKey;
        this.value = value;
        this.assetId = assetId;
        this.nonce = nonce;
        this.noteSecret = noteSecret;
    }
    toBuffer() {
        return Buffer.concat([
            bigint_buffer_1.toBufferBE(this.value, 32),
            serialize_1.numToUInt32BE(this.assetId),
            serialize_1.numToUInt32BE(this.nonce),
            this.ownerPubKey.toBuffer(),
            this.noteSecret,
        ]);
    }
    getViewingKey(ephPrivKey, grumpkin) {
        const noteBuf = Buffer.concat([bigint_buffer_1.toBufferBE(this.value, 32), serialize_1.numToUInt32BE(this.assetId), serialize_1.numToUInt32BE(this.nonce)]);
        return viewing_key_1.ViewingKey.createFromEphPriv(noteBuf, this.ownerPubKey, ephPrivKey, grumpkin);
    }
    static createFromEphPriv(ownerPubKey, value, assetId, nonce, ephPrivKey, grumpkin, noteVersion = 1) {
        const noteSecret = derive_note_secret_1.deriveNoteSecret(ownerPubKey, ephPrivKey, grumpkin, noteVersion);
        return new TreeNote(ownerPubKey, value, assetId, nonce, noteSecret);
    }
    static createFromEphPub(ownerPubKey, value, assetId, nonce, ephPubKey, ownerPrivKey, grumpkin, noteVersion = 1) {
        const noteSecret = derive_note_secret_1.deriveNoteSecret(ephPubKey, ownerPrivKey, grumpkin, noteVersion);
        return new TreeNote(ownerPubKey, value, assetId, nonce, noteSecret);
    }
    static recover({ noteBuf, noteSecret }, ownerPubKey) {
        const value = bigint_buffer_1.toBigIntBE(noteBuf.slice(0, 32));
        const assetId = noteBuf.readUInt32BE(32);
        const nonce = noteBuf.readUInt32BE(36);
        return new TreeNote(ownerPubKey, value, assetId, nonce, noteSecret);
    }
}
exports.TreeNote = TreeNote;
TreeNote.EMPTY = new TreeNote(address_1.GrumpkinAddress.one(), BigInt(0), 0, 0, Buffer.alloc(32));
