// @aztec/sdk/web_sdk/eth_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthProvider = exports.chainIdToNetwork = exports.networks = exports.EthProviderEvent = void 0;
const tslib_1 = require("tslib");
const events_1 = require("events");
const debug_1 = tslib_1.__importDefault(require("debug"));
const address_1 = require("@aztec/barretenberg/address");
const providers_1 = require("@ethersproject/providers");
const debug = debug_1.default('bb:eth_provider');
var EthProviderEvent;
(function (EthProviderEvent) {
    EthProviderEvent["UPDATED_NETWORK"] = "UPDATED_NETWORK";
    EthProviderEvent["UPDATED_ACCOUNT"] = "UPDATED_ACCOUNT";
})(EthProviderEvent = exports.EthProviderEvent || (exports.EthProviderEvent = {}));
exports.networks = [
    { chainId: 1, network: 'mainnet' },
    { chainId: 3, network: 'ropsten' },
    { chainId: 4, network: 'rinkeby' },
    { chainId: 5, network: 'goerli' },
    { chainId: 42, network: 'kovan' },
    { chainId: 1337, network: 'ganache' },
];
exports.chainIdToNetwork = (chainId) => {
    var _a;
    return ((_a = exports.networks.find(network => network.chainId === chainId)) === null || _a === void 0 ? void 0 : _a.network) || 'unknown';
};
class EthProvider extends events_1.EventEmitter {
    constructor(provider) {
        super();
        this.provider = provider;
        this.chainId = -1;
        this.accounts = [];
        this.updateNetwork = async () => {
            const ethersProvider = new providers_1.Web3Provider(this.provider);
            this.chainId = (await ethersProvider.getNetwork()).chainId;
            const network = exports.chainIdToNetwork(this.chainId);
            debug(`chainId set to: ${this.chainId}`);
            this.emit(EthProviderEvent.UPDATED_NETWORK, network, this.chainId);
        };
        this.updateAccounts = (accounts) => {
            this.accounts = accounts.map(address_1.EthAddress.fromString);
            this.emit(EthProviderEvent.UPDATED_ACCOUNT, this.accounts[0]);
        };
    }
    async init() {
        const ethersProvider = new providers_1.Web3Provider(this.provider);
        await this.provider.request({ method: 'eth_requestAccounts' });
        await this.updateNetwork();
        const accounts = await ethersProvider.listAccounts();
        this.updateAccounts(accounts);
        this.provider.on('chainChanged', this.updateNetwork);
        this.provider.on('accountsChanged', this.updateAccounts);
        return this.getAccount();
    }
    destroy() {
        this.provider.removeListener('chainChanged', this.updateNetwork);
        this.provider.removeListener('accountsChanged', this.updateAccounts);
        this.removeAllListeners();
    }
    getChainId() {
        return this.chainId;
    }
    getNetwork() {
        return exports.chainIdToNetwork(this.getChainId());
    }
    getAccounts() {
        return this.accounts;
    }
    getAccount() {
        return this.accounts[0];
    }
}
exports.EthProvider = EthProvider;
