// @aztec/sdk/web_sdk/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSdk = exports.AppEvent = exports.AppInitAction = exports.AppInitState = void 0;
const tslib_1 = require("tslib");
const service_1 = require("@aztec/barretenberg/service");
const debug_1 = tslib_1.__importDefault(require("debug"));
const events_1 = require("events");
const ethereum_sdk_1 = require("../ethereum_sdk");
const sdk_1 = require("../sdk");
const eth_provider_1 = require("./eth_provider");
const debug = debug_1.default('bb:websdk');
var AppInitState;
(function (AppInitState) {
    AppInitState["UNINITIALIZED"] = "UNINITIALIZED";
    AppInitState["INITIALIZING"] = "INITIALIZING";
    AppInitState["INITIALIZED"] = "INITIALIZED";
})(AppInitState = exports.AppInitState || (exports.AppInitState = {}));
var AppInitAction;
(function (AppInitAction) {
    AppInitAction["LINK_PROVIDER_ACCOUNT"] = "LINK_PROVIDER_ACCOUNT";
    AppInitAction["AWAITING_PROVIDER_SIGNATURE"] = "AWAITING_PROVIDER_SIGNATURE";
    AppInitAction["AWAITING_PERMISSION_TO_LINK"] = "AWAITING_PERMISSION_TO_LINK";
    AppInitAction["CHANGE_NETWORK"] = "CHANGE_NETWORK";
})(AppInitAction = exports.AppInitAction || (exports.AppInitAction = {}));
var AppEvent;
(function (AppEvent) {
    AppEvent["UPDATED_INIT_STATE"] = "APPEVENT_UPDATED_INIT_STATE";
})(AppEvent = exports.AppEvent || (exports.AppEvent = {}));
/**
 * Simplifies integration of the CoreSdk with a provider such as MetaMask.
 * The event stream will always be ordered like, but may not always include, the following:
 *
 * Initialization starts:
 * UPDATED_INIT_STATE => INITIALIZING, LINK_PROVIDER_ACCOUNT
 * UPDATED_INIT_STATE => INITIALIZING, CHANGE_NETWORK
 * UPDATED_INIT_STATE => INITIALIZING, "info message 1"
 * UPDATED_INIT_STATE => INITIALIZING, "info message 2"
 * UPDATED_INIT_STATE => INITIALIZING, AWAITING_PROVIDER_SIGNATURE (user accepts, otherwise destroy)
 * UPDATED_INIT_STATE => INITIALIZED, address 1
 * UPDATED_INIT_STATE => INITIALIZING, AWAITING_PERMISSION_TO_LINK (user changes to unlinked account)
 * UPDATED_INIT_STATE => INITIALIZING, AWAITING_PROVIDER_SIGNATURE (user rejects)
 * UPDATED_INIT_STATE => INITIALIZING, AWAITING_PERMISSION_TO_LINK
 * UPDATED_INIT_STATE => INITIALIZING, AWAITING_PROVIDER_SIGNATURE (user accepts)
 * UPDATED_INIT_STATE => INITIALIZED, address 2
 * UPDATED_INIT_STATE => INITIALIZED, address 1 (user changes to linked account)
 * UPDATED_INIT_STATE => DESTROYED
 */
class WebSdk extends events_1.EventEmitter {
    constructor(ethereumProvider) {
        super();
        this.ethereumProvider = ethereumProvider;
        this.initStatus = { initState: AppInitState.UNINITIALIZED };
        this.accountChanged = async (account) => {
            if (!account) {
                this.destroy();
                return;
            }
            this.user = await this.sdk.getUser(account);
            if (!this.user) {
                // We are initializing until the account is added to sdk.
                debug(`provider emitted account changed to ${account}, requesting permission to link...`);
                this.updateInitStatus(AppInitState.INITIALIZING, AppInitAction.AWAITING_PERMISSION_TO_LINK);
            }
            else {
                this.updateInitStatus(AppInitState.INITIALIZED);
            }
        };
        this.networkChanged = () => {
            if (!this.isCorrectNetwork()) {
                this.destroy();
            }
        };
        this.linkAccount = async () => {
            const account = this.ethProvider.getAccount();
            if (!account) {
                this.destroy();
                return;
            }
            this.user = await this.sdk.getUser(account);
            if (!this.user) {
                // Call to addUser will attempt to perform a signature. Alert client.
                this.updateInitStatus(AppInitState.INITIALIZING, AppInitAction.AWAITING_PROVIDER_SIGNATURE);
                try {
                    this.user = await this.sdk.addUser(account);
                }
                catch (e) {
                    debug(e);
                    this.updateInitStatus(AppInitState.INITIALIZING, AppInitAction.AWAITING_PERMISSION_TO_LINK);
                    return;
                }
            }
            this.updateInitStatus(AppInitState.INITIALIZED);
        };
    }
    async init(serverUrl, sdkOptions = {}) {
        if (sdkOptions.debug) {
            debug_1.default.enable('bb:*');
        }
        this.ethereumProvider.autoRefreshOnNetworkChange = false;
        debug('initializing app...');
        try {
            const { chainId: rollupProviderChainId } = await service_1.getBlockchainStatus(serverUrl);
            this.updateInitStatus(AppInitState.INITIALIZING, AppInitAction.LINK_PROVIDER_ACCOUNT);
            this.ethProvider = new eth_provider_1.EthProvider(this.ethereumProvider);
            await this.ethProvider.init();
            // If our network doesn't match that of the rollup provider, request it be changed until it does.
            this.initStatus.network = eth_provider_1.chainIdToNetwork(rollupProviderChainId);
            if (rollupProviderChainId !== this.ethProvider.getChainId()) {
                this.updateInitStatus(AppInitState.INITIALIZING, AppInitAction.CHANGE_NETWORK);
                while (rollupProviderChainId !== this.ethProvider.getChainId()) {
                    await new Promise(resolve => setTimeout(resolve, 500));
                }
            }
            this.sdk = await ethereum_sdk_1.createEthSdk(this.ethereumProvider, serverUrl, sdkOptions);
            // Forward all sdk events. This allows subscribing to the events on the App, before we have called init().
            for (const e in sdk_1.SdkEvent) {
                const event = sdk_1.SdkEvent[e];
                this.sdk.on(event, (...args) => this.emit(event, ...args));
            }
            await this.sdk.init();
            // Link account. Will be INITIALZED once complete.
            await this.initLinkAccount();
            // Handle account changes.
            this.ethProvider.on(eth_provider_1.EthProviderEvent.UPDATED_ACCOUNT, this.accountChanged);
            // Ensure we're still on correct network, and attach handler.
            // Any network changes at this point result in destruction.
            this.networkChanged();
            this.ethProvider.on(eth_provider_1.EthProviderEvent.UPDATED_NETWORK, this.networkChanged);
            debug('initialization complete.');
        }
        catch (err) {
            this.destroy();
            throw err;
        }
    }
    updateInitStatus(initState, initAction) {
        const previous = this.initStatus;
        this.initStatus = {
            ...this.initStatus,
            initState,
            initAction,
        };
        // debug(`state ${previous.initState} -> ${initState}: ${initAction}`);
        this.emit(AppEvent.UPDATED_INIT_STATE, { ...this.initStatus }, previous);
    }
    async initLinkAccount() {
        const account = this.ethProvider.getAccount();
        if (!account) {
            throw new Error('Account access withdrawn.');
        }
        this.user = await this.sdk.getUser(account);
        if (!this.user) {
            // Call to addUser will attempt to perform a signature. Alert client.
            this.updateInitStatus(AppInitState.INITIALIZING, AppInitAction.AWAITING_PROVIDER_SIGNATURE);
            try {
                this.user = await this.sdk.addUser(account);
            }
            catch (e) {
                debug(e);
                throw new Error('Account link rejected.');
            }
        }
        this.updateInitStatus(AppInitState.INITIALIZED);
    }
    async loadLatestAccount() {
        this.user = await this.sdk.getUser(this.ethProvider.getAccount());
    }
    async destroy() {
        var _a, _b;
        debug('destroying app...');
        await ((_a = this.sdk) === null || _a === void 0 ? void 0 : _a.destroy());
        (_b = this.ethProvider) === null || _b === void 0 ? void 0 : _b.destroy();
        this.user === undefined;
        this.updateInitStatus(AppInitState.UNINITIALIZED);
    }
    getSdk() {
        return this.sdk;
    }
    isInitialized() {
        return this.getInitStatus().initState === AppInitState.INITIALIZED;
    }
    isCorrectNetwork() {
        const { chainId } = this.sdk.getLocalStatus();
        return this.ethProvider.getChainId() === chainId;
    }
    getInitStatus() {
        return this.initStatus;
    }
    getUser() {
        return this.user;
    }
    getAddress() {
        return this.ethProvider.getAccount();
    }
}
exports.WebSdk = WebSdk;
