// @aztec/blockchain/provider/wallet_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletProvider = void 0;
const ethers_1 = require("ethers");
const address_1 = require("@aztec/barretenberg/address");
const ethers_adapter_1 = require("./ethers_adapter");
const providers_1 = require("@ethersproject/providers");
/**
 * Given an EIP1193 provider, wraps it, and provides the ability to add local accounts.
 */
class WalletProvider {
    constructor(provider) {
        this.provider = provider;
        this.accounts = [];
    }
    static fromHost(ethereumHost) {
        const ethersProvider = new providers_1.JsonRpcProvider(ethereumHost);
        return new WalletProvider(new ethers_adapter_1.EthersAdapter(ethersProvider));
    }
    addAccount(privateKey) {
        return this.addEthersWallet(new ethers_1.Wallet(privateKey));
    }
    addEthersWallet(wallet) {
        this.accounts.push(wallet);
        return address_1.EthAddress.fromString(wallet.address);
    }
    getAccounts() {
        return this.accounts.map(a => address_1.EthAddress.fromString(a.address));
    }
    getAccount(account) {
        return address_1.EthAddress.fromString(this.accounts[account].address);
    }
    getPrivateKey(account) {
        return Buffer.from(this.accounts[account].privateKey.slice(2), 'hex');
    }
    getPrivateKeyForAddress(account) {
        const wallet = this.accounts.find(w => account.equals(address_1.EthAddress.fromString(w.address)));
        return wallet ? Buffer.from(wallet.privateKey.slice(2), 'hex') : undefined;
    }
    async request(args) {
        switch (args.method) {
            case 'eth_accounts':
                return this.accounts.length ? this.accounts.map(a => a.address) : await this.provider.request(args);
            case 'eth_sign':
                return await this.sign(args);
            case 'eth_signTypedData_v4':
                return this.signTypedData(args);
            case 'eth_signTransaction':
                return this.signTransaction(args);
            case 'eth_sendTransaction':
                return this.sendTransaction(args);
            default: {
                return await this.provider.request(args);
            }
        }
    }
    async sign(args) {
        const [from, message] = args.params;
        const account = this.accounts.find(a => a.address.toLowerCase() === from);
        if (account) {
            return await account.signMessage(Buffer.from(message.slice(2), 'hex'));
        }
        return await this.provider.request(args);
    }
    async signTypedData(args) {
        const [from, data] = args.params;
        const { types, domain, message } = JSON.parse(data);
        const account = this.accounts.find(a => a.address.toLowerCase() === from);
        if (account) {
            delete types.EIP712Domain;
            return await account._signTypedData(domain, types, message);
        }
        return this.provider.request(args);
    }
    /**
     * Given a tx in Eth Json Rpc format, convert to ethers format and give to account to sign.
     * Populate any missing fields.
     */
    async signTxLocally(tx, account) {
        const gasLimit = tx.gas || 7000000;
        const gasPrice = tx.gasPrice || (await this.provider.request({ method: 'eth_gasPrice' }));
        const value = tx.value || 0;
        const chainId = +(await this.provider.request({ method: 'eth_chainId' }));
        const nonce = tx.nonce || (await this.provider.request({ method: 'eth_getTransactionCount', params: [tx.from, 'latest'] }));
        const toSign = {
            chainId,
            from: tx.from,
            to: tx.to,
            data: tx.data,
            gasLimit,
            gasPrice,
            value,
            nonce,
        };
        return await account.signTransaction(toSign);
    }
    async signTransaction(args) {
        const tx = args.params[0];
        const account = this.accounts.find(a => a.address.toLowerCase() === tx.from.toLowerCase());
        if (account) {
            return this.signTxLocally(tx, account);
        }
        return this.provider.request(args);
    }
    async sendTransaction(args) {
        const tx = args.params[0];
        const account = this.accounts.find(a => a.address.toLowerCase() === tx.from.toLowerCase());
        if (account) {
            const result = await this.signTxLocally(tx, account);
            return this.provider.request({ method: 'eth_sendRawTransaction', params: [result] });
        }
        return this.provider.request(args);
    }
    on(notification, listener) {
        return this.provider.on(notification, listener);
    }
    removeListener(notification, listener) {
        return this.provider.removeListener(notification, listener);
    }
}
exports.WalletProvider = WalletProvider;
