// @aztec/blockchain/provider/web3_adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Web3Adapter = void 0;
/**
 * Adapts a legacy web3 provider into an EIP1193 compatible provider for injecting into the sdk.
 */
class Web3Adapter {
    constructor(provider) {
        this.provider = provider;
        this.id = 0;
    }
    request(args) {
        return new Promise((resolve, reject) => {
            const payload = { jsonrpc: '2.0', id: this.id++, method: args.method, params: args.params || [] };
            this.provider.send(payload, (err, response) => {
                if (err) {
                    return reject(err);
                }
                if (!response) {
                    return reject(new Error('No response.'));
                }
                resolve(response.result);
            });
        });
    }
    on() {
        throw new Error('Events not supported.');
    }
    removeListener() {
        throw new Error('Events not supported.');
    }
}
exports.Web3Adapter = Web3Adapter;
