// @aztec/barretenberg/client_proofs/join_split_proof/join_split_tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JoinSplitTx = void 0;
const bigint_buffer_1 = require("../../bigint_buffer");
const serialize_1 = require("../../serialize");
class JoinSplitTx {
    constructor(publicInput, publicOutput, assetId, numInputNotes, inputNoteIndices, merkleRoot, inputNotePaths, inputNotes, outputNotes, claimNote, accountPrivateKey, accountAliasId, accountIndex, accountPath, signingPubKey, signature, inputOwner, outputOwner) {
        this.publicInput = publicInput;
        this.publicOutput = publicOutput;
        this.assetId = assetId;
        this.numInputNotes = numInputNotes;
        this.inputNoteIndices = inputNoteIndices;
        this.merkleRoot = merkleRoot;
        this.inputNotePaths = inputNotePaths;
        this.inputNotes = inputNotes;
        this.outputNotes = outputNotes;
        this.claimNote = claimNote;
        this.accountPrivateKey = accountPrivateKey;
        this.accountAliasId = accountAliasId;
        this.accountIndex = accountIndex;
        this.accountPath = accountPath;
        this.signingPubKey = signingPubKey;
        this.signature = signature;
        this.inputOwner = inputOwner;
        this.outputOwner = outputOwner;
    }
    toBuffer() {
        return Buffer.concat([
            bigint_buffer_1.toBufferBE(this.publicInput, 32),
            bigint_buffer_1.toBufferBE(this.publicOutput, 32),
            serialize_1.numToUInt32BE(this.assetId),
            serialize_1.numToUInt32BE(this.numInputNotes),
            serialize_1.numToUInt32BE(this.inputNoteIndices[0]),
            serialize_1.numToUInt32BE(this.inputNoteIndices[1]),
            this.merkleRoot,
            this.inputNotePaths[0].toBuffer(),
            this.inputNotePaths[1].toBuffer(),
            this.inputNotes[0].toBuffer(),
            this.inputNotes[1].toBuffer(),
            this.outputNotes[0].toBuffer(),
            this.outputNotes[1].toBuffer(),
            this.claimNote.toBuffer(),
            this.accountPrivateKey,
            this.accountAliasId.aliasHash.toBuffer32(),
            serialize_1.numToUInt32BE(this.accountAliasId.nonce),
            serialize_1.numToUInt32BE(this.accountIndex),
            this.accountPath.toBuffer(),
            this.signingPubKey.toBuffer(),
            this.signature.toBuffer(),
            this.inputOwner.toBuffer32(),
            this.outputOwner.toBuffer32(),
        ]);
    }
}
exports.JoinSplitTx = JoinSplitTx;
