// @aztec/barretenberg/rollup_provider/server_rollup_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerRollupProvider = void 0;
const iso_fetch_1 = require("../iso_fetch");
const block_source_1 = require("../block_source");
const tx_hash_1 = require("../tx_hash");
const blockchain_1 = require("../blockchain");
class ServerRollupProvider extends block_source_1.ServerBlockSource {
    constructor(baseUrl, pollInterval = 10000) {
        super(baseUrl, pollInterval);
    }
    async sendProof({ proofData, viewingKeys, depositSignature, ...rest }) {
        const url = new URL(`${this.baseUrl}/tx`);
        const data = {
            proofData: proofData.toString('hex'),
            viewingKeys: viewingKeys.map(v => v.toString()),
            depositSignature: depositSignature ? depositSignature.toString('hex') : undefined,
            ...rest,
        };
        const response = await iso_fetch_1.fetch(url.toString(), { method: 'POST', body: JSON.stringify(data) }).catch(() => undefined);
        if (!response) {
            throw new Error('Failed to contact rollup provider.');
        }
        if (response.status === 400) {
            const body = await response.json();
            throw new Error(body.error);
        }
        if (response.status !== 200) {
            throw new Error(`Bad response code ${response.status}.`);
        }
        const body = await response.json();
        return tx_hash_1.TxHash.fromString(body.txHash);
    }
    async getStatus() {
        const url = new URL(`${this.baseUrl}/status`);
        const response = await iso_fetch_1.fetch(url.toString()).catch(() => undefined);
        if (!response) {
            throw new Error('Failed to contact rollup provider.');
        }
        try {
            const { txFees, blockchainStatus, nextPublishTime, ...rest } = await response.json();
            return {
                ...rest,
                blockchainStatus: blockchain_1.blockchainStatusFromJson(blockchainStatus),
                txFees: txFees.map(({ feeConstants, baseFeeQuotes }) => ({
                    feeConstants: feeConstants.map(r => BigInt(r)),
                    baseFeeQuotes: baseFeeQuotes.map(({ fee, time }) => ({
                        time,
                        fee: BigInt(fee),
                    })),
                })),
                nextPublishTime: new Date(nextPublishTime),
            };
        }
        catch (err) {
            throw new Error(`Bad response from: ${url}`);
        }
    }
    async getPendingNoteNullifiers() {
        const url = new URL(`${this.baseUrl}/get-pending-note-nullifiers`);
        const response = await iso_fetch_1.fetch(url.toString());
        if (response.status !== 200) {
            throw new Error(`Bad response code ${response.status}.`);
        }
        const nullifiers = (await response.json());
        return nullifiers.map(n => Buffer.from(n, 'hex'));
    }
}
exports.ServerRollupProvider = ServerRollupProvider;
