// @aztec/sdk/database/sql_database/transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bridgeIdTransformer = exports.viewingKeyTransformer = exports.ethAddressTransformer = exports.txHashTransformer = exports.accountIdTransformer = exports.accountAliasIdTransformer = exports.aliasHashTransformer = exports.grumpkinAddressTransformer = exports.bigintTransformer = void 0;
const account_id_1 = require("@aztec/barretenberg/account_id");
const address_1 = require("@aztec/barretenberg/address");
const bridge_id_1 = require("@aztec/barretenberg/bridge_id");
const tx_hash_1 = require("@aztec/barretenberg/tx_hash");
const viewing_key_1 = require("@aztec/barretenberg/viewing_key");
const user_1 = require("../../user");
exports.bigintTransformer = {
    to: (entityValue) => (entityValue !== undefined ? `${entityValue}` : ''),
    from: (dbValue) => (dbValue ? BigInt(dbValue) : undefined),
};
exports.grumpkinAddressTransformer = {
    to: (entityValue) => entityValue === null || entityValue === void 0 ? void 0 : entityValue.toBuffer(),
    from: (dbValue) => (dbValue ? new address_1.GrumpkinAddress(dbValue) : undefined),
};
exports.aliasHashTransformer = {
    to: (entityValue) => entityValue === null || entityValue === void 0 ? void 0 : entityValue.toBuffer(),
    from: (dbValue) => (dbValue ? new account_id_1.AliasHash(dbValue) : undefined),
};
exports.accountAliasIdTransformer = {
    to: (entityValue) => entityValue === null || entityValue === void 0 ? void 0 : entityValue.toBuffer(),
    from: (dbValue) => (dbValue ? user_1.AccountAliasId.fromBuffer(dbValue) : undefined),
};
exports.accountIdTransformer = {
    to: (entityValue) => entityValue === null || entityValue === void 0 ? void 0 : entityValue.toBuffer(),
    from: (dbValue) => (dbValue ? user_1.AccountId.fromBuffer(dbValue) : undefined),
};
exports.txHashTransformer = {
    to: (entityValue) => entityValue === null || entityValue === void 0 ? void 0 : entityValue.toBuffer(),
    from: (dbValue) => (dbValue ? new tx_hash_1.TxHash(dbValue) : undefined),
};
exports.ethAddressTransformer = {
    to: (entityValue) => entityValue === null || entityValue === void 0 ? void 0 : entityValue.toBuffer(),
    from: (dbValue) => (dbValue ? new address_1.EthAddress(dbValue) : undefined),
};
exports.viewingKeyTransformer = {
    to: (entityValue) => entityValue === null || entityValue === void 0 ? void 0 : entityValue.toBuffer(),
    from: (dbValue) => (dbValue ? new viewing_key_1.ViewingKey(dbValue) : undefined),
};
exports.bridgeIdTransformer = {
    to: (entityValue) => entityValue === null || entityValue === void 0 ? void 0 : entityValue.toBuffer(),
    from: (dbValue) => (dbValue ? bridge_id_1.BridgeId.fromBuffer(dbValue) : undefined),
};
