// @aztec/barretenberg/blockchain/blockchain_status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blockchainStatusFromJson = exports.blockchainStatusToJson = exports.TxType = void 0;
const address_1 = require("../address");
var TxType;
(function (TxType) {
    TxType[TxType["DEPOSIT"] = 0] = "DEPOSIT";
    TxType[TxType["TRANSFER"] = 1] = "TRANSFER";
    TxType[TxType["WITHDRAW_TO_WALLET"] = 2] = "WITHDRAW_TO_WALLET";
    TxType[TxType["WITHDRAW_TO_CONTRACT"] = 3] = "WITHDRAW_TO_CONTRACT";
    TxType[TxType["ACCOUNT"] = 4] = "ACCOUNT";
    TxType[TxType["DEFI_DEPOSIT"] = 5] = "DEFI_DEPOSIT";
    TxType[TxType["DEFI_CLAIM"] = 6] = "DEFI_CLAIM";
})(TxType = exports.TxType || (exports.TxType = {}));
function blockchainStatusToJson(status) {
    return {
        ...status,
        rollupContractAddress: status.rollupContractAddress.toString(),
        feeDistributorContractAddress: status.feeDistributorContractAddress.toString(),
        dataRoot: status.dataRoot.toString('hex'),
        nullRoot: status.nullRoot.toString('hex'),
        rootRoot: status.rootRoot.toString('hex'),
        defiRoot: status.defiRoot.toString('hex'),
        defiInteractionHash: status.defiInteractionHash.toString('hex'),
        totalDeposited: status.totalDeposited.map(f => f.toString()),
        totalWithdrawn: status.totalWithdrawn.map(f => f.toString()),
        totalPendingDeposit: status.totalPendingDeposit.map(f => f.toString()),
        totalFees: status.totalFees.map(f => f.toString()),
        feeDistributorBalance: status.feeDistributorBalance.map(f => f.toString()),
        assets: status.assets.map(a => ({
            ...a,
            address: a.address.toString(),
        })),
    };
}
exports.blockchainStatusToJson = blockchainStatusToJson;
function blockchainStatusFromJson(json) {
    return {
        ...json,
        rollupContractAddress: address_1.EthAddress.fromString(json.rollupContractAddress),
        feeDistributorContractAddress: address_1.EthAddress.fromString(json.feeDistributorContractAddress),
        dataRoot: Buffer.from(json.dataRoot, 'hex'),
        nullRoot: Buffer.from(json.nullRoot, 'hex'),
        rootRoot: Buffer.from(json.rootRoot, 'hex'),
        defiRoot: Buffer.from(json.defiRoot, 'hex'),
        defiInteractionHash: Buffer.from(json.defiInteractionHash, 'hex'),
        totalDeposited: json.totalDeposited.map(f => BigInt(f)),
        totalWithdrawn: json.totalWithdrawn.map(f => BigInt(f)),
        totalPendingDeposit: json.totalPendingDeposit.map(f => BigInt(f)),
        totalFees: json.totalFees.map(f => BigInt(f)),
        feeDistributorBalance: json.feeDistributorBalance.map(f => BigInt(f)),
        assets: json.assets.map(a => ({
            ...a,
            address: address_1.EthAddress.fromString(a.address),
        })),
    };
}
exports.blockchainStatusFromJson = blockchainStatusFromJson;
