// @aztec/barretenberg/bridge_id/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgeId = void 0;
const address_1 = require("../address");
const bigint_buffer_1 = require("../bigint_buffer");
const randomInt = (to = 2 ** 26 - 1) => Math.floor(Math.random() * (to + 1));
const getNumber = (val, offset, size) => Number((val >> BigInt(offset)) & ((BigInt(1) << BigInt(size)) - BigInt(1)));
class BridgeId {
    constructor(address, numOutputAssets, inputAssetId, outputAssetIdA, outputAssetIdB) {
        this.address = address;
        this.numOutputAssets = numOutputAssets;
        this.inputAssetId = inputAssetId;
        this.outputAssetIdA = outputAssetIdA;
        this.outputAssetIdB = outputAssetIdB;
    }
    static random() {
        return new BridgeId(address_1.EthAddress.randomAddress(), 1 + randomInt(1), randomInt(), randomInt(), randomInt());
    }
    static fromBigInt(val) {
        return new BridgeId(new address_1.EthAddress(bigint_buffer_1.toBufferBE(val & ((BigInt(1) << BigInt(this.ADDRESS_BIT_LEN)) - BigInt(1)), 32)), getNumber(val, this.ADDRESS_BIT_LEN, this.NUM_OUTPUT_NOTES_LEN), getNumber(val, this.ADDRESS_BIT_LEN + this.NUM_OUTPUT_NOTES_LEN, this.INPUT_ASSET_ID_LEN), getNumber(val, this.ADDRESS_BIT_LEN + this.NUM_OUTPUT_NOTES_LEN + this.INPUT_ASSET_ID_LEN, this.OUTPUT_A_ASSET_ID_LEN), getNumber(val, this.ADDRESS_BIT_LEN + this.NUM_OUTPUT_NOTES_LEN + this.INPUT_ASSET_ID_LEN + this.OUTPUT_A_ASSET_ID_LEN, this.OUTPUT_B_ASSET_ID_LEN));
    }
    static fromBuffer(buf) {
        if (buf.length !== 32) {
            throw new Error('Invalid buffer.');
        }
        return BridgeId.fromBigInt(bigint_buffer_1.toBigIntBE(buf));
    }
    static fromString(str) {
        return BridgeId.fromBuffer(Buffer.from(str.replace(/^0x/i, ''), 'hex'));
    }
    toBigInt() {
        return (BigInt(this.address.toString()) +
            (BigInt(this.numOutputAssets) << BigInt(160)) +
            (BigInt(this.inputAssetId) << BigInt(162)) +
            (BigInt(this.outputAssetIdA) << BigInt(192)) +
            (BigInt(this.outputAssetIdB) << BigInt(222)));
    }
    toBuffer() {
        return bigint_buffer_1.toBufferBE(this.toBigInt(), 32);
    }
    toString() {
        return `0x${this.toBuffer().toString('hex')}`;
    }
    equals(id) {
        return id.toBuffer().equals(this.toBuffer());
    }
}
exports.BridgeId = BridgeId;
BridgeId.ZERO = new BridgeId(address_1.EthAddress.ZERO, 0, 0, 0, 0);
BridgeId.LENGTH = 32;
BridgeId.ADDRESS_BIT_LEN = 160;
BridgeId.NUM_OUTPUT_NOTES_LEN = 2;
BridgeId.INPUT_ASSET_ID_LEN = 30;
BridgeId.OUTPUT_A_ASSET_ID_LEN = 30;
BridgeId.OUTPUT_B_ASSET_ID_LEN = 30;
