// @aztec/sdk/core_sdk/create_sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSdk = void 0;
const tslib_1 = require("tslib");
const rollup_provider_1 = require("@aztec/barretenberg/rollup_provider");
const broadcast_channel_1 = require("broadcast-channel");
const debug_1 = tslib_1.__importDefault(require("debug"));
const detect_node_1 = tslib_1.__importDefault(require("detect-node"));
const fs_1 = require("fs");
const levelup_1 = tslib_1.__importDefault(require("levelup"));
const typeorm_1 = require("typeorm");
const database_1 = require("../database");
const sdk_1 = require("../sdk");
const user_1 = require("../user");
const core_sdk_1 = require("./core_sdk");
const debug = debug_1.default('bb:create_sdk');
function getLevelDb(dbPath = 'data') {
    if (detect_node_1.default) {
        if (dbPath === ':memory:') {
            // eslint-disable-next-line
            return levelup_1.default(require('memdown')());
        }
        else {
            fs_1.mkdirSync(dbPath, { recursive: true });
            // eslint-disable-next-line
            return levelup_1.default(require('leveldown')(`${dbPath}/aztec2-sdk.db`));
        }
    }
    else {
        // eslint-disable-next-line
        return levelup_1.default(require('level-js')('aztec2-sdk'));
    }
}
async function getDb(dbPath = 'data') {
    if (detect_node_1.default) {
        const config = database_1.getOrmConfig(dbPath);
        const connection = await typeorm_1.createConnection(config);
        return new database_1.SQLDatabase(connection);
    }
    else {
        return new database_1.DexieDatabase();
    }
}
async function sdkFactory(hostStr, options) {
    if (options.debug) {
        debug_1.default.enable('bb:*');
    }
    const host = new URL(hostStr);
    const leveldb = getLevelDb(options.dbPath);
    const db = await getDb(options.dbPath);
    await db.init();
    if (options.clearDb) {
        await leveldb.clear();
        await db.clear();
    }
    const rollupProvider = new rollup_provider_1.ServerRollupProvider(host);
    return new core_sdk_1.CoreSdk(leveldb, db, rollupProvider, options);
}
/**
 * Construct an SDK instance. If passed the `syncInstances` option, will bind a channel to various events to
 * share events and synchronise instances. Only one instance will be the "leader" and that instance will receive
 * blocks from the block source and update the (shared) world state.
 */
async function createSdk(hostStr, options = {}) {
    options = { syncInstances: true, saveProvingKey: true, ...options };
    const sdk = await sdkFactory(hostStr, options);
    if (!options.syncInstances) {
        // We're not going to sync across multiple instances. We should start recieving blocks once initialized.
        sdk.on(sdk_1.SdkEvent.UPDATED_INIT_STATE, state => {
            if (state === sdk_1.SdkInitState.INITIALIZED) {
                sdk.startReceivingBlocks();
            }
        });
    }
    else {
        // We're going to sync across multiple instances.
        const channel = new broadcast_channel_1.BroadcastChannel('aztec-sdk');
        // If this instance becomes the leader at any time, and we are initialized, start recieving blocks.
        const elector = broadcast_channel_1.createLeaderElection(channel, { responseTime: 2000 });
        elector.awaitLeadership().then(() => {
            if (sdk.getLocalStatus().initState === sdk_1.SdkInitState.INITIALIZED) {
                sdk.startReceivingBlocks();
            }
            else {
                debug('elected leader, will wait until initialized to process blocks.');
            }
        });
        sdk.on(sdk_1.SdkEvent.UPDATED_INIT_STATE, state => {
            if (state === sdk_1.SdkInitState.INITIALIZED && elector.isLeader) {
                sdk.startReceivingBlocks();
            }
            else if (state === sdk_1.SdkInitState.DESTROYED) {
                channel.close();
            }
        });
        sdk.on(core_sdk_1.CoreSdkEvent.UPDATED_WORLD_STATE, () => channel.postMessage({ name: core_sdk_1.CoreSdkEvent.UPDATED_WORLD_STATE }));
        sdk.on(core_sdk_1.CoreSdkEvent.UPDATED_USERS, () => channel.postMessage({ name: core_sdk_1.CoreSdkEvent.UPDATED_USERS }));
        sdk.on(core_sdk_1.CoreSdkEvent.UPDATED_USER_STATE, (userId, balanceAfter, diff, assetId) => channel.postMessage({
            name: core_sdk_1.CoreSdkEvent.UPDATED_USER_STATE,
            userId: userId.toString(),
            balanceAfter: balanceAfter === null || balanceAfter === void 0 ? void 0 : balanceAfter.toString(),
            diff: diff === null || diff === void 0 ? void 0 : diff.toString(),
            assetId,
        }));
        channel.onmessage = msg => {
            if (sdk.getLocalStatus().initState !== sdk_1.SdkInitState.INITIALIZED) {
                return;
            }
            switch (msg.name) {
                case core_sdk_1.CoreSdkEvent.UPDATED_WORLD_STATE:
                    sdk.notifyWorldStateUpdated();
                    break;
                case core_sdk_1.CoreSdkEvent.UPDATED_USERS:
                    sdk.initUserStates();
                    break;
                case core_sdk_1.CoreSdkEvent.UPDATED_USER_STATE:
                    sdk.notifyUserStateUpdated(user_1.AccountId.fromString(msg.userId), msg.balanceAfter ? BigInt(msg.balanceAfter) : undefined, msg.diff ? BigInt(msg.diff) : undefined, msg.assetId);
                    break;
            }
        };
    }
    return sdk;
}
exports.createSdk = createSdk;
