// @aztec/barretenberg/serialize/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeArrayFromVector = exports.serializeBufferArrayToVector = exports.deserializeField = exports.deserializeBigInt = exports.deserializeUInt32 = exports.deserializeBufferFromVector = exports.serializeBufferToVector = exports.numToUInt8 = exports.numToUInt32BE = void 0;
const bigint_buffer_1 = require("../bigint_buffer");
// For serializing numbers to 32 bit big-endian form.
function numToUInt32BE(n, bufferSize = 4) {
    const buf = Buffer.alloc(bufferSize);
    buf.writeUInt32BE(n, bufferSize - 4);
    return buf;
}
exports.numToUInt32BE = numToUInt32BE;
// For serializing numbers to 32 bit big-endian form.
function numToUInt8(n) {
    const bufferSize = 1;
    const buf = Buffer.alloc(bufferSize);
    buf.writeUInt8(n, 0);
    return buf;
}
exports.numToUInt8 = numToUInt8;
// For serializing a buffer as a vector.
function serializeBufferToVector(buf) {
    const lengthBuf = Buffer.alloc(4);
    lengthBuf.writeUInt32BE(buf.length, 0);
    return Buffer.concat([lengthBuf, buf]);
}
exports.serializeBufferToVector = serializeBufferToVector;
function deserializeBufferFromVector(vector, offset = 0) {
    const length = vector.readUInt32BE(offset);
    const adv = 4 + length;
    return { elem: vector.slice(offset + 4, offset + adv), adv };
}
exports.deserializeBufferFromVector = deserializeBufferFromVector;
function deserializeUInt32(buf, offset = 0) {
    const adv = 4;
    return { elem: buf.readUInt32BE(offset), adv };
}
exports.deserializeUInt32 = deserializeUInt32;
function deserializeBigInt(buf, offset = 0) {
    const adv = 32;
    return { elem: bigint_buffer_1.toBigIntBE(buf.slice(offset, 32)), adv };
}
exports.deserializeBigInt = deserializeBigInt;
function deserializeField(buf, offset = 0) {
    const adv = 32;
    return { elem: buf.slice(offset, offset + adv), adv };
}
exports.deserializeField = deserializeField;
// For serializing an array of fixed length elements.
function serializeBufferArrayToVector(arr) {
    const lengthBuf = Buffer.alloc(4);
    lengthBuf.writeUInt32BE(arr.length, 0);
    return Buffer.concat([lengthBuf, ...arr]);
}
exports.serializeBufferArrayToVector = serializeBufferArrayToVector;
function deserializeArrayFromVector(deserialize, vector, offset = 0) {
    let startAt = offset;
    const size = vector.readUInt32BE(startAt);
    startAt += 4;
    const arr = [];
    for (let i = 0; i < size; ++i) {
        const { elem, adv } = deserialize(vector, startAt);
        startAt += adv;
        arr.push(elem);
    }
    return { elem: arr, adv: startAt - offset };
}
exports.deserializeArrayFromVector = deserializeArrayFromVector;
