// @aztec/barretenberg/rollup_proof/inner_proof.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InnerProofData = void 0;
const client_proofs_1 = require("../client_proofs");
const serialize_1 = require("../serialize");
class InnerProofData {
    constructor(proofId, publicInput, publicOutput, assetId, noteCommitment1, noteCommitment2, nullifier1, nullifier2, inputOwner, outputOwner) {
        this.proofId = proofId;
        this.publicInput = publicInput;
        this.publicOutput = publicOutput;
        this.assetId = assetId;
        this.noteCommitment1 = noteCommitment1;
        this.noteCommitment2 = noteCommitment2;
        this.nullifier1 = nullifier1;
        this.nullifier2 = nullifier2;
        this.inputOwner = inputOwner;
        this.outputOwner = outputOwner;
        this.txId = client_proofs_1.createTxId(this.toBuffer());
    }
    getDepositSigningData() {
        return this.toBuffer();
    }
    toBuffer() {
        return Buffer.concat([
            serialize_1.numToUInt32BE(this.proofId, 32),
            this.publicInput,
            this.publicOutput,
            this.assetId,
            this.noteCommitment1,
            this.noteCommitment2,
            this.nullifier1,
            this.nullifier2,
            this.inputOwner,
            this.outputOwner,
        ]);
    }
    isPadding() {
        return this.nullifier1.equals(Buffer.alloc(32, 0));
    }
    static fromBuffer(innerPublicInputs) {
        const proofId = innerPublicInputs.readUInt32BE(0 * 32 + 28);
        const publicInput = innerPublicInputs.slice(1 * 32, 1 * 32 + 32);
        const publicOutput = innerPublicInputs.slice(2 * 32, 2 * 32 + 32);
        const assetId = innerPublicInputs.slice(3 * 32, 3 * 32 + 32);
        const noteCommitment1 = innerPublicInputs.slice(4 * 32, 4 * 32 + 32);
        const noteCommitment2 = innerPublicInputs.slice(5 * 32, 5 * 32 + 32);
        const nullifier1 = innerPublicInputs.slice(6 * 32, 6 * 32 + 32);
        const nullifier2 = innerPublicInputs.slice(7 * 32, 7 * 32 + 32);
        const inputOwner = innerPublicInputs.slice(8 * 32, 8 * 32 + 32);
        const outputOwner = innerPublicInputs.slice(9 * 32, 9 * 32 + 32);
        return new InnerProofData(proofId, publicInput, publicOutput, assetId, noteCommitment1, noteCommitment2, nullifier1, nullifier2, inputOwner, outputOwner);
    }
}
exports.InnerProofData = InnerProofData;
InnerProofData.NUM_PUBLIC_INPUTS = 10;
InnerProofData.LENGTH = InnerProofData.NUM_PUBLIC_INPUTS * 32;
