// @aztec/blockchain/client_ethereum_blockchain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientEthereumBlockchain = void 0;
const providers_1 = require("@ethersproject/providers");
const address_1 = require("@aztec/barretenberg/address");
const asset_1 = require("./asset");
const hash_data_1 = require("./hash_data");
const rollup_processor_1 = require("./rollup_processor");
class ClientEthereumBlockchain {
    constructor(rollupContractAddress, assetInfos, ethereumProvider, minConfirmation = 1) {
        this.ethereumProvider = ethereumProvider;
        this.minConfirmation = minConfirmation;
        this.provider = new providers_1.Web3Provider(this.ethereumProvider);
        this.rollupProcessor = new rollup_processor_1.RollupProcessor(rollupContractAddress, this.provider);
        this.assets = assetInfos.map(info => info.address.equals(address_1.EthAddress.ZERO)
            ? new asset_1.EthAsset(this.provider)
            : new asset_1.TokenAsset(this.provider, info, minConfirmation));
    }
    getAsset(assetId) {
        return this.assets[assetId];
    }
    async getUserPendingDeposit(assetId, account) {
        return this.rollupProcessor.getUserPendingDeposit(assetId, account);
    }
    async getUserProofApprovalStatus(account, signingData) {
        const proofHash = hash_data_1.hashData(signingData);
        return this.rollupProcessor.getUserProofApprovalStatus(account, proofHash);
    }
    async depositPendingFunds(assetId, amount, from, permitArgs, provider) {
        return this.rollupProcessor.depositPendingFunds(assetId, amount, permitArgs, this.getEthSigner(from, provider));
    }
    async approveProof(account, signingData, provider) {
        const proofHash = hash_data_1.hashData(signingData);
        return this.rollupProcessor.approveProof(proofHash, this.getEthSigner(account, provider));
    }
    /**
     * Wait for given transaction to be mined, and return receipt.
     */
    async getTransactionReceipt(txHash, interval = 1, timeout = 300, minConfirmation = this.minConfirmation) {
        const started = Date.now();
        while (true) {
            if (timeout && Date.now() - started > timeout * 1000) {
                throw new Error(`Timeout awaiting tx confirmation: ${txHash}`);
            }
            const txReceipt = await this.provider.getTransactionReceipt(txHash.toString());
            if (!minConfirmation || (txReceipt && txReceipt.confirmations >= minConfirmation)) {
                return txReceipt
                    ? { status: !!txReceipt.status, blockNum: txReceipt.blockNumber }
                    : { status: false, blockNum: 0 };
            }
            await new Promise(resolve => setTimeout(resolve, interval * 1000));
        }
    }
    async isContract(address) {
        return (await this.provider.getCode(address.toString())) !== '0x';
    }
    getEthSigner(address, provider) {
        return (provider ? new providers_1.Web3Provider(provider) : this.provider).getSigner(address.toString());
    }
}
exports.ClientEthereumBlockchain = ClientEthereumBlockchain;
