// @aztec/blockchain/create_permit_data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPermitData = void 0;
const types = {
    Permit: [
        {
            name: 'owner',
            type: 'address',
        },
        {
            name: 'spender',
            type: 'address',
        },
        {
            name: 'value',
            type: 'uint256',
        },
        {
            name: 'nonce',
            type: 'uint256',
        },
        {
            name: 'deadline',
            type: 'uint256',
        },
    ],
};
const getDomain = (name, chainId, verifyingContract) => ({
    name,
    version: '1',
    chainId,
    verifyingContract,
});
const getMessage = (owner, spender, value, nonce, deadline) => ({
    owner: owner.toString(),
    spender: spender.toString(),
    value,
    nonce,
    deadline,
});
function createPermitData(name, owner, spender, value, nonce, deadline, chainId, verifyingContract) {
    const domain = getDomain(name, chainId, verifyingContract.toString());
    const message = getMessage(owner, spender, value, nonce, deadline);
    return { types, domain, message };
}
exports.createPermitData = createPermitData;
