// @aztec/blockchain/ethereum_blockchain.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumBlockchain = void 0;
const bridge_id_1 = require("@aztec/barretenberg/bridge_id");
const debug_1 = __importDefault(require("debug"));
const events_1 = require("events");
const contracts_1 = require("./contracts");
const hash_data_1 = require("./hash_data");
const validate_signature_1 = require("./validate_signature");
class EthereumBlockchain extends events_1.EventEmitter {
    constructor(config, contracts) {
        super();
        this.config = config;
        this.contracts = contracts;
        this.running = false;
        this.latestEthBlock = -1;
        this.latestRollupId = -1;
        this.debug = config.console === false ? debug_1.default('bb:ethereum_blockchain') : console.log;
    }
    static async new(config, rollupContractAddress, priceFeedContractAddresses, provider) {
        const confirmations = config.minConfirmation || EthereumBlockchain.DEFAULT_MIN_CONFIRMATIONS;
        const contracts = new contracts_1.Contracts(rollupContractAddress, priceFeedContractAddresses, provider, confirmations);
        await contracts.init();
        const eb = new EthereumBlockchain(config, contracts);
        await eb.init();
        return eb;
    }
    async init() {
        await this.initStatus();
        this.debug(`Ethereum blockchain initialized with assets: ${this.status.assets.map(a => a.symbol)}`);
    }
    /**
     * Start polling for RollupProcessed events.
     * All historical blocks will have been emitted before this function returns.
     */
    async start(fromRollup = 0) {
        this.debug(`Ethereum blockchain starting from rollup: ${fromRollup}`);
        const getBlocks = async (fromRollup) => {
            while (true) {
                try {
                    return await this.getBlocks(fromRollup);
                }
                catch (err) {
                    this.debug(`getBlocks failed, will retry: ${err.message}`);
                    await new Promise(resolve => setTimeout(resolve, 5000));
                }
            }
        };
        const emitBlocks = async () => {
            const latestBlock = await this.contracts.getBlockNumber().catch(err => {
                this.debug(`getBlockNumber failed: ${err.code}`);
                return this.latestEthBlock;
            });
            if (latestBlock === this.latestEthBlock) {
                return;
            }
            this.latestEthBlock = latestBlock;
            await this.updatePerBlockState();
            const blocks = await getBlocks(fromRollup);
            if (blocks.length) {
                await this.updatePerRollupState();
            }
            for (const block of blocks) {
                this.debug(`Block received: ${block.rollupId}`);
                this.latestRollupId = block.rollupId;
                this.emit('block', block);
                fromRollup = block.rollupId + 1;
            }
        };
        // We must have emitted all historical blocks before returning.
        await emitBlocks();
        // After which, we asynchronously kick off a polling loop for the latest blocks.
        this.running = true;
        (async () => {
            while (this.running) {
                await new Promise(resolve => setTimeout(resolve, this.config.pollInterval || 10000));
                await emitBlocks().catch(this.debug);
            }
        })();
    }
    /**
     * Stop polling for RollupProcessed events
     */
    async stop() {
        this.running = false;
        this.removeAllListeners();
    }
    /**
     * Get the status of the rollup contract
     */
    async getBlockchainStatus(refresh = false) {
        if (refresh) {
            await this.initStatus();
        }
        return this.status;
    }
    async initStatus() {
        await this.updatePerRollupState();
        await this.updatePerBlockState();
        const { chainId } = await this.contracts.getNetwork();
        const assets = this.contracts.getAssets().map(a => a.getStaticInfo());
        this.status = {
            ...this.status,
            chainId,
            rollupContractAddress: this.contracts.getRollupContractAddress(),
            feeDistributorContractAddress: this.contracts.getFeeDistributorContractAddress(),
            assets,
        };
    }
    async updatePerRollupState() {
        this.status = {
            ...this.status,
            ...(await this.contracts.getPerRollupState()),
        };
    }
    async updatePerBlockState() {
        this.status = {
            ...this.status,
            ...(await this.contracts.getPerBlockState()),
        };
    }
    getLatestRollupId() {
        return this.latestRollupId;
    }
    async getUserPendingDeposit(assetId, account) {
        return this.contracts.getUserPendingDeposit(assetId, account);
    }
    async getUserProofApprovalStatus(account, signingData) {
        const proofHash = hash_data_1.hashData(signingData);
        return this.contracts.getUserProofApprovalStatus(account, proofHash);
    }
    async setSupportedAsset(assetAddress, supportsPermit, signingAddress) {
        return this.contracts.setSupportedAsset(assetAddress, supportsPermit, signingAddress);
    }
    async createRollupProofTx(proofData, signatures, viewingKeys, providerSignature, providerAddress, feeReceiver, feeLimit) {
        return await this.contracts.createRollupProofTx(proofData, signatures, viewingKeys, providerSignature, providerAddress, feeReceiver, feeLimit);
    }
    async createEscapeHatchProofTx(proofData, viewingKeys, depositSignature, signingAddress) {
        return await this.contracts.createEscapeHatchProofTx(proofData, viewingKeys, depositSignature ? [depositSignature] : [], signingAddress);
    }
    sendTx(tx, options = {}) {
        options = { ...options, gasLimit: options.gasLimit || this.config.gasLimit };
        return this.contracts.sendTx(tx, options);
    }
    getRequiredConfirmations() {
        const { escapeOpen, numEscapeBlocksRemaining } = this.status;
        const { minConfirmation = EthereumBlockchain.DEFAULT_MIN_CONFIRMATIONS, minConfirmationEHW = EthereumBlockchain.DEFAULT_MIN_CONFIRMATIONS_EHW, } = this.config;
        return escapeOpen || numEscapeBlocksRemaining <= minConfirmationEHW ? minConfirmationEHW : minConfirmation;
    }
    /**
     * Get all created rollup blocks from `rollupId`.
     */
    async getBlocks(rollupId) {
        const minConfirmations = this.getRequiredConfirmations();
        return await this.contracts.getRollupBlocksFrom(rollupId, minConfirmations);
    }
    /**
     * Wait for given transaction to be mined, and return receipt.
     */
    async getTransactionReceipt(txHash) {
        const confs = this.config.minConfirmation || EthereumBlockchain.DEFAULT_MIN_CONFIRMATIONS;
        this.debug(`Getting tx receipt for ${txHash}... (${confs} confirmations)`);
        let txReceipt = await this.contracts.getTransactionReceipt(txHash);
        while (!txReceipt || txReceipt.confirmations < confs) {
            await new Promise(resolve => setTimeout(resolve, 1000));
            txReceipt = await this.contracts.getTransactionReceipt(txHash);
        }
        return { status: !!txReceipt.status, blockNum: txReceipt.blockNumber };
    }
    async getTransactionReceiptSafe(txHash) {
        const confs = this.getRequiredConfirmations();
        this.debug(`Getting tx receipt for ${txHash} (${confs} confs)...`);
        let txReceipt = await this.contracts.getTransactionReceipt(txHash);
        while (!txReceipt || txReceipt.confirmations < confs) {
            await new Promise(resolve => setTimeout(resolve, 1000));
            txReceipt = await this.contracts.getTransactionReceipt(txHash);
        }
        return { status: !!txReceipt.status, blockNum: txReceipt.blockNumber };
    }
    /**
     * Validate locally that a signature was produced by a publicOwner
     */
    validateSignature(publicOwner, signature, signingData) {
        return validate_signature_1.validateSignature(publicOwner, signature, signingData);
    }
    async signMessage(message, address) {
        return this.contracts.signMessage(message, address);
    }
    async signTypedData(data, address) {
        return this.contracts.signTypedData(data, address);
    }
    getAsset(assetId) {
        return this.contracts.getAsset(assetId);
    }
    async getAssetPrice(assetId) {
        return this.contracts.getAssetPrice(assetId);
    }
    getPriceFeed(assetId) {
        return this.contracts.getPriceFeed(assetId);
    }
    getGasPriceFeed() {
        return this.contracts.getGasPriceFeed();
    }
    async getBridgeId(address) {
        try {
            return await this.contracts.getBridgeId(address);
        }
        catch (e) {
            this.debug(e);
            return bridge_id_1.BridgeId.ZERO;
        }
    }
    async isContract(address) {
        return this.contracts.isContract(address);
    }
    async getGasPrice() {
        return this.contracts.getGasPrice();
    }
    async estimateGas(data) {
        return this.contracts.estimateGas(data);
    }
}
exports.EthereumBlockchain = EthereumBlockchain;
EthereumBlockchain.DEFAULT_MIN_CONFIRMATIONS = 3;
EthereumBlockchain.DEFAULT_MIN_CONFIRMATIONS_EHW = 12;
