// @aztec/barretenberg/client_proofs/account_proof/account_tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountTx = void 0;
const serialize_1 = require("../../serialize");
class AccountTx {
    constructor(merkleRoot, accountPublicKey, newAccountPublicKey, numNewKeys, newSigningPubKey1, newSigningPubKey2, accountAliasId, migrate, gibberish, accountIndex, accountPath, signingPubKey, signature) {
        this.merkleRoot = merkleRoot;
        this.accountPublicKey = accountPublicKey;
        this.newAccountPublicKey = newAccountPublicKey;
        this.numNewKeys = numNewKeys;
        this.newSigningPubKey1 = newSigningPubKey1;
        this.newSigningPubKey2 = newSigningPubKey2;
        this.accountAliasId = accountAliasId;
        this.migrate = migrate;
        this.gibberish = gibberish;
        this.accountIndex = accountIndex;
        this.accountPath = accountPath;
        this.signingPubKey = signingPubKey;
        this.signature = signature;
        if (gibberish.length !== 32) {
            throw new Error('gibberish should be 32-byte long.');
        }
    }
    toBuffer() {
        return Buffer.concat([
            this.merkleRoot,
            this.accountPublicKey.toBuffer(),
            this.newAccountPublicKey.toBuffer(),
            serialize_1.numToUInt32BE(this.numNewKeys),
            this.newSigningPubKey1.toBuffer(),
            this.newSigningPubKey2.toBuffer(),
            this.accountAliasId.aliasHash.toBuffer32(),
            serialize_1.numToUInt32BE(this.accountAliasId.nonce),
            Buffer.from([this.migrate ? 1 : 0]),
            this.gibberish,
            serialize_1.numToUInt32BE(this.accountIndex),
            this.accountPath.toBuffer(),
            this.signingPubKey.toBuffer(),
            this.signature.toBuffer(),
        ]);
    }
}
exports.AccountTx = AccountTx;
