// @aztec/barretenberg/address/grumpkin_address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrumpkinAddress = void 0;
const crypto_1 = require("crypto");
const grumpkin_1 = require("../ecc/grumpkin");
class GrumpkinAddress {
    constructor(buffer) {
        this.buffer = buffer;
        if (buffer.length !== 64) {
            throw new Error('Invalid address buffer.');
        }
    }
    static isAddress(address) {
        return /^(0x|0X)?[0-9a-fA-F]{128}$/.test(address);
    }
    static fromString(address) {
        if (!GrumpkinAddress.isAddress(address)) {
            throw new Error(`Invalid address string: ${address}`);
        }
        return new GrumpkinAddress(Buffer.from(address.replace(/^0x/i, ''), 'hex'));
    }
    /**
     * NOT a valid address! Do not use in proofs.
     */
    static randomAddress() {
        return new GrumpkinAddress(crypto_1.randomBytes(64));
    }
    /**
     * A valid address (is a point on the curve).
     */
    static one() {
        return new GrumpkinAddress(grumpkin_1.Grumpkin.one);
    }
    equals(rhs) {
        return this.buffer.equals(rhs.toBuffer());
    }
    toBuffer() {
        return this.buffer;
    }
    x() {
        return this.buffer.slice(0, 32);
    }
    y() {
        return this.buffer.slice(32);
    }
    toString() {
        return `0x${this.buffer.toString('hex')}`;
    }
}
exports.GrumpkinAddress = GrumpkinAddress;
GrumpkinAddress.ZERO = new GrumpkinAddress(Buffer.alloc(64));
