// @aztec/sdk/sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SdkInitState = exports.SdkEvent = void 0;
var SdkEvent;
(function (SdkEvent) {
    // Initialization state changes.
    SdkEvent["UPDATED_INIT_STATE"] = "SDKEVENT_UPDATED_INIT_STATE";
    // The set of users has changed.
    SdkEvent["UPDATED_USERS"] = "SDKEVENT_UPDATED_USERS";
    // A users state has changed.
    SdkEvent["UPDATED_USER_STATE"] = "SDKEVENT_UPDATED_USER_STATE";
    // The world state has updated. Used for displaying sync progress.
    SdkEvent["UPDATED_WORLD_STATE"] = "SDKEVENT_UPDATED_WORLD_STATE";
    // Log messages for long running operations.
    SdkEvent["LOG"] = "SDKEVENT_LOG";
})(SdkEvent = exports.SdkEvent || (exports.SdkEvent = {}));
var SdkInitState;
(function (SdkInitState) {
    SdkInitState["UNINITIALIZED"] = "UNINITIALIZED";
    SdkInitState["INITIALIZING"] = "INITIALIZING";
    SdkInitState["INITIALIZED"] = "INITIALIZED";
    SdkInitState["DESTROYED"] = "DESTROYED";
})(SdkInitState = exports.SdkInitState || (exports.SdkInitState = {}));
