// @aztec/sdk/wallet_sdk/wallet_sdk_user_asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletSdkUserAsset = void 0;
class WalletSdkUserAsset {
    constructor(userId, assetId, sdk) {
        this.userId = userId;
        this.assetId = assetId;
        this.sdk = sdk;
    }
    info() {
        return this.sdk.getAssetInfo(this.assetId);
    }
    balance() {
        return this.sdk.getBalance(this.assetId, this.userId);
    }
    async publicBalance(account) {
        return this.sdk.getPublicBalance(this.assetId, account);
    }
    async publicAllowance(account) {
        return this.sdk.getPublicAllowance(this.assetId, account);
    }
    async pendingDeposit(account) {
        return this.sdk.getUserPendingDeposit(this.assetId, account);
    }
    async getMaxSpendableValue() {
        return this.sdk.getMaxSpendableValue(this.assetId, this.userId);
    }
    async mint(value, account) {
        return this.sdk.mint(this.assetId, value, account);
    }
    async approve(value, account) {
        return this.sdk.approve(this.assetId, value, account);
    }
    async depositFundsToContract(from, value, permitArgs) {
        return this.sdk.depositFundsToContract(this.assetId, from, value, permitArgs);
    }
    async createDepositProof(value, fee, signer, from) {
        return this.sdk.createDepositProof(this.assetId, from, this.userId, value, fee, signer);
    }
    async createWithdrawProof(value, fee, signer, to) {
        return this.sdk.createWithdrawProof(this.assetId, this.userId, value, fee, signer, to);
    }
    async createTransferProof(value, fee, signer, to) {
        return this.sdk.createTransferProof(this.assetId, this.userId, value, fee, signer, to);
    }
    fromBaseUnits(value, precision) {
        return this.sdk.fromBaseUnits(this.assetId, value, precision);
    }
    toBaseUnits(value) {
        return this.sdk.toBaseUnits(this.assetId, value);
    }
    async getFee(txType) {
        return this.sdk.getFee(this.assetId, txType);
    }
}
exports.WalletSdkUserAsset = WalletSdkUserAsset;
