// @aztec/sdk/wallet_sdk/wallet_sdk_user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletSdkUser = void 0;
const wallet_sdk_user_asset_1 = require("./wallet_sdk_user_asset");
class WalletSdkUser {
    constructor(id, sdk) {
        this.id = id;
        this.sdk = sdk;
    }
    async awaitSynchronised() {
        return this.sdk.awaitUserSynchronised(this.id);
    }
    async createAccount(alias, newSigningPublicKey, recoveryPublicKey) {
        return this.sdk.createAccount(this.id, alias, newSigningPublicKey, recoveryPublicKey);
    }
    async recoverAccount(recoveryPayload) {
        return this.sdk.recoverAccount(recoveryPayload);
    }
    async migrateAccount(signer, newSigningPublicKey, recoveryPublicKey, newAccountPrivateKey) {
        return this.sdk.migrateAccount(this.id, signer, newSigningPublicKey, recoveryPublicKey, newAccountPrivateKey);
    }
    async addSigningKeys(signer, signingPublicKey1, signingPublicKey2) {
        return this.sdk.addSigningKeys(this.id, signer, signingPublicKey1, signingPublicKey2);
    }
    async getSigningKeys() {
        return this.sdk.getSigningKeys(this.id);
    }
    getUserData() {
        return this.sdk.getUserData(this.id);
    }
    async getJoinSplitTxs() {
        return this.sdk.getJoinSplitTxs(this.id);
    }
    async getAccountTxs() {
        return this.sdk.getAccountTxs(this.id);
    }
    async getNotes() {
        return this.sdk.getNotes(this.id);
    }
    getAsset(assetId) {
        return new wallet_sdk_user_asset_1.WalletSdkUserAsset(this.id, assetId, this.sdk);
    }
    getSigner() {
        return this.sdk.createSchnorrSigner(this.getUserData().privateKey);
    }
}
exports.WalletSdkUser = WalletSdkUser;
