// @aztec/sdk/proofs/proof_output.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefiProofOutput = exports.AccountProofOutput = exports.JoinSplitProofOutput = void 0;
const address_1 = require("@aztec/barretenberg/address");
const client_proofs_1 = require("@aztec/barretenberg/client_proofs");
const tx_hash_1 = require("@aztec/barretenberg/tx_hash");
const user_1 = require("../user");
const user_tx_1 = require("../user_tx");
class JoinSplitProofOutput {
    constructor(tx, proofData, viewingKeys, signingData) {
        this.tx = tx;
        this.proofData = proofData;
        this.viewingKeys = viewingKeys;
        this.signingData = signingData;
    }
}
exports.JoinSplitProofOutput = JoinSplitProofOutput;
class AccountProofOutput {
    constructor(tx, proofData) {
        this.tx = tx;
        this.proofData = proofData;
        this.viewingKeys = [];
    }
    static fromBuffer(buf) {
        const [migratedBuf, rawProofData] = [buf.slice(0, 1), buf.slice(1)];
        const proofData = new client_proofs_1.ProofData(rawProofData);
        const accountProof = new client_proofs_1.AccountProofData(proofData);
        const publicKey = new address_1.GrumpkinAddress(accountProof.publicKey);
        const { nonce, aliasHash } = accountProof.accountAliasId;
        const tx = new user_tx_1.UserAccountTx(new tx_hash_1.TxHash(proofData.txId), new user_1.AccountId(publicKey, nonce), aliasHash, proofData.inputOwner, proofData.outputOwner, !migratedBuf.equals(Buffer.alloc(1)), new Date());
        return new AccountProofOutput(tx, rawProofData);
    }
    toBuffer() {
        return Buffer.concat([Buffer.from([+this.tx.migrated]), this.proofData]);
    }
}
exports.AccountProofOutput = AccountProofOutput;
class DefiProofOutput {
    constructor(tx, proofData, viewingKeys) {
        this.tx = tx;
        this.proofData = proofData;
        this.viewingKeys = viewingKeys;
    }
}
exports.DefiProofOutput = DefiProofOutput;
