// @aztec/sdk/proofs/account_proof_creator/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountProofCreator = void 0;
const tslib_1 = require("tslib");
const address_1 = require("@aztec/barretenberg/address");
const client_proofs_1 = require("@aztec/barretenberg/client_proofs");
const crypto_1 = require("crypto");
const debug_1 = tslib_1.__importDefault(require("debug"));
const user_1 = require("../../user");
const debug = debug_1.default('bb:account_proof');
class AccountProofCreator {
    constructor(accountProver, worldState, pedersen) {
        this.accountProver = accountProver;
        this.worldState = worldState;
        this.pedersen = pedersen;
    }
    async createAccountTx(signer, aliasHash, nonce, migrate, accountPublicKey, newAccountPublicKey, newSigningPubKey1, newSigningPubKey2, accountIndex = 0) {
        const merkleRoot = this.worldState.getRoot();
        const numNewKeys = [newSigningPubKey1, newSigningPubKey2].filter(k => !!k).length;
        const signingPubKey = signer.getPublicKey();
        const accountPath = await this.worldState.getHashPath(accountIndex);
        const accountAliasId = new user_1.AccountAliasId(aliasHash, nonce);
        const gibberish = crypto_1.randomBytes(32);
        const sigMsg = client_proofs_1.computeAccountProofSigningData(accountAliasId, accountPublicKey, newAccountPublicKey || accountPublicKey, newSigningPubKey1 || address_1.GrumpkinAddress.ZERO, newSigningPubKey2 || address_1.GrumpkinAddress.ZERO, this.pedersen);
        const signature = await signer.signMessage(sigMsg);
        return new client_proofs_1.AccountTx(merkleRoot, accountPublicKey, newAccountPublicKey || accountPublicKey, numNewKeys, newSigningPubKey1 || address_1.GrumpkinAddress.ZERO, newSigningPubKey2 || address_1.GrumpkinAddress.ZERO, accountAliasId, migrate, gibberish, accountIndex, accountPath, signingPubKey, signature);
    }
    async createProof(signer, aliasHash, nonce, migrate, accountPublicKey, newAccountPublicKey, newSigningPubKey1, newSigningPubKey2, accountIndex) {
        const tx = await this.createAccountTx(signer, aliasHash, nonce, migrate, accountPublicKey, newAccountPublicKey, newSigningPubKey1, newSigningPubKey2, accountIndex);
        debug('creating proof...');
        const start = new Date().getTime();
        const proofData = await this.accountProver.createAccountProof(tx);
        debug(`created proof: ${new Date().getTime() - start}ms`);
        debug(`proof size: ${proofData.length}`);
        return proofData;
    }
}
exports.AccountProofCreator = AccountProofCreator;
