// @aztec/barretenberg/client_proofs/proof_data/join_split_proof_data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JoinSplitProofData = void 0;
const address_1 = require("../../address");
const bigint_buffer_1 = require("../../bigint_buffer");
const create_tx_id_1 = require("./create_tx_id");
class JoinSplitProofData {
    constructor(proofData) {
        this.proofData = proofData;
        this.assetId = this.proofData.assetId.readUInt32BE(28);
        this.publicInput = bigint_buffer_1.toBigIntBE(this.proofData.publicInput);
        this.publicOutput = bigint_buffer_1.toBigIntBE(this.proofData.publicOutput);
        this.inputOwner = new address_1.EthAddress(this.proofData.inputOwner.slice(12));
        this.outputOwner = new address_1.EthAddress(this.proofData.outputOwner.slice(12));
        /**
         * TODO: Get rid of this in favor of just signing tx id.
         * The data we sign over for authorizing deposits, consists of the data that is published on chain.
         * This excludes the last two fields, the noteTreeRoot and the txFee.
         */
        this.depositSigningData = this.proofData.rawProofData.slice(0, create_tx_id_1.LENGTH_PROOF_HEADER_INPUTS);
    }
}
exports.JoinSplitProofData = JoinSplitProofData;
