// @aztec/sdk/ethereum_sdk/ethereum_sdk_user_asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumSdkUserAsset = void 0;
class EthereumSdkUserAsset {
    constructor(address, accountId, assetId, sdk) {
        this.address = address;
        this.accountId = accountId;
        this.assetId = assetId;
        this.sdk = sdk;
    }
    getInfo() {
        return this.sdk.getAssetInfo(this.assetId);
    }
    async publicBalance() {
        return this.sdk.getPublicBalance(this.assetId, this.address);
    }
    async publicAllowance() {
        return this.sdk.getPublicAllowance(this.assetId, this.address);
    }
    async pendingDeposit() {
        return this.sdk.getUserPendingDeposit(this.assetId, this.address);
    }
    balance() {
        return this.sdk.getBalance(this.assetId, this.accountId);
    }
    async getMaxSpendableValue() {
        return this.sdk.getMaxSpendableValue(this.assetId, this.accountId);
    }
    async mint(value) {
        return this.sdk.mint(this.assetId, value, this.address);
    }
    async approve(value) {
        return this.sdk.approve(this.assetId, value, this.address);
    }
    async depositFundsToContract(value, permitArgs) {
        return this.sdk.depositFundsToContract(this.assetId, this.address, value, permitArgs);
    }
    async createDepositProof(value, fee) {
        return this.sdk.createDepositProof(this.assetId, this.address, this.accountId, value, fee);
    }
    async createWithdrawProof(value, fee) {
        return this.sdk.createWithdrawProof(this.assetId, this.accountId, this.address, value, fee);
    }
    async createTransferProof(value, fee, to) {
        return this.sdk.createTransferProof(this.assetId, this.accountId, to, value, fee);
    }
    async signProof(proofOutput) {
        return this.sdk.signProof(proofOutput, this.address);
    }
    fromBaseUnits(value, precision) {
        return this.sdk.fromBaseUnits(this.assetId, value, precision);
    }
    toBaseUnits(value) {
        return this.sdk.toBaseUnits(this.assetId, value);
    }
    async getFee(txType) {
        return this.sdk.getFee(this.assetId, txType);
    }
}
exports.EthereumSdkUserAsset = EthereumSdkUserAsset;
