// @aztec/sdk/ethereum_sdk/ethereum_sdk_user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumSdkUser = void 0;
const ethereum_sdk_user_asset_1 = require("./ethereum_sdk_user_asset");
class EthereumSdkUser {
    constructor(address, accountId, sdk) {
        this.address = address;
        this.accountId = accountId;
        this.sdk = sdk;
    }
    isSynching() {
        return this.sdk.isUserSynching(this.accountId);
    }
    async awaitSynchronised() {
        return this.sdk.awaitUserSynchronised(this.accountId);
    }
    createAccount(alias, newSigningPublicKey, recoveryPublicKey) {
        return this.sdk.createAccount(this.accountId, alias, newSigningPublicKey, recoveryPublicKey);
    }
    getUserData() {
        return this.sdk.getUserData(this.accountId);
    }
    remove() {
        return this.sdk.removeUser(this.address, this.accountId);
    }
    async getJoinSplitTxs() {
        return this.sdk.getJoinSplitTxs(this.accountId);
    }
    async getAccountTxs() {
        return this.sdk.getAccountTxs(this.accountId);
    }
    async getNotes() {
        return this.sdk.getNotes(this.accountId);
    }
    getAsset(assetId) {
        return new ethereum_sdk_user_asset_1.EthereumSdkUserAsset(this.address, this.accountId, assetId, this.sdk);
    }
}
exports.EthereumSdkUser = EthereumSdkUser;
