// @aztec/barretenberg/block_source/server_block_source.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerBlockSource = void 0;
const events_1 = require("events");
const note_algorithms_1 = require("../note_algorithms");
const iso_fetch_1 = require("../iso_fetch");
const tx_hash_1 = require("../tx_hash");
// const debug = createDebug('bb:server_block_source');
const toBlock = (block) => ({
    ...block,
    txHash: tx_hash_1.TxHash.fromString(block.txHash),
    rollupProofData: Buffer.from(block.rollupProofData, 'hex'),
    viewingKeysData: Buffer.from(block.viewingKeysData, 'hex'),
    interactionResult: block.interactionResult.map(r => note_algorithms_1.DefiInteractionNote.fromBuffer(Buffer.from(r, 'hex'))),
    created: new Date(block.created),
    gasPrice: BigInt(block.gasPrice),
});
class ServerBlockSource extends events_1.EventEmitter {
    constructor(baseUrl, pollInterval = 10000) {
        super();
        this.pollInterval = pollInterval;
        this.running = false;
        this.runningPromise = Promise.resolve();
        this.interruptPromise = Promise.resolve();
        this.interruptResolve = () => { };
        this.latestRollupId = -1;
        this.baseUrl = baseUrl.toString().replace(/\/$/, '');
    }
    getLatestRollupId() {
        return this.latestRollupId;
    }
    async start(from = 0) {
        this.running = true;
        this.interruptPromise = new Promise(resolve => (this.interruptResolve = resolve));
        const emitBlocks = async () => {
            try {
                const blocks = await this.getBlocks(from);
                for (const block of blocks) {
                    this.emit('block', block);
                    from = block.rollupId + 1;
                }
            }
            catch (err) {
                // debug(err);
            }
        };
        await emitBlocks();
        const poll = async () => {
            while (this.running) {
                await emitBlocks();
                await this.sleepOrInterrupted(this.pollInterval);
            }
        };
        this.runningPromise = poll();
    }
    stop() {
        this.running = false;
        this.interruptResolve();
        return this.runningPromise;
    }
    async awaitSucceed(fn) {
        while (true) {
            try {
                const response = await fn();
                if (response.status !== 200) {
                    throw new Error(`Bad status code: ${response.status}`);
                }
                return response;
            }
            catch (err) {
                console.log(err.message);
                await new Promise(resolve => setTimeout(resolve, 1000));
            }
        }
    }
    async getBlocks(from) {
        const url = new URL(`${this.baseUrl}/get-blocks`);
        url.searchParams.append('from', from.toString());
        const response = await this.awaitSucceed(() => iso_fetch_1.fetch(url.toString()));
        const result = (await response.json());
        this.latestRollupId = result.latestRollupId;
        return result.blocks.map(toBlock);
    }
    async sleepOrInterrupted(ms) {
        let timeout;
        const promise = new Promise(resolve => (timeout = setTimeout(resolve, ms)));
        await Promise.race([promise, this.interruptPromise]);
        clearTimeout(timeout);
    }
}
exports.ServerBlockSource = ServerBlockSource;
