// @aztec/sdk/proofs/join_split_proof_creator/join_split_tx_factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JoinSplitTxFactory = void 0;
const address_1 = require("@aztec/barretenberg/address");
const client_proofs_1 = require("@aztec/barretenberg/client_proofs");
const note_algorithms_1 = require("@aztec/barretenberg/note_algorithms");
const viewing_key_1 = require("@aztec/barretenberg/viewing_key");
const user_1 = require("../../user");
class JoinSplitTxFactory {
    constructor(worldState, grumpkin, pedersen, noteAlgos, db) {
        this.worldState = worldState;
        this.grumpkin = grumpkin;
        this.pedersen = pedersen;
        this.noteAlgos = noteAlgos;
        this.db = db;
    }
    async createJoinSplitTx(userState, publicInput, publicOutput, privateInput, recipientPrivateOutput, senderPrivateOutput, defiDepositValue, assetId, signer, newNoteOwner, inputOwnerAddress, outputOwnerAddress, bridgeId) {
        const isDefiBridge = defiDepositValue > BigInt(0);
        const { id, aliasHash, publicKey, nonce } = userState.getUser();
        const accountIndex = nonce !== 0 ? await this.db.getUserSigningKeyIndex(id, signer.getPublicKey()) : 0;
        if (accountIndex === undefined) {
            throw new Error('Unknown signing key.');
        }
        const accountAliasId = aliasHash ? new user_1.AccountAliasId(aliasHash, nonce) : user_1.AccountAliasId.random();
        const accountPath = await this.worldState.getHashPath(accountIndex);
        const signingPubKey = signer.getPublicKey();
        const notes = privateInput ? await userState.pickNotes(assetId, privateInput) : [];
        if (!notes) {
            throw new Error(`Failed to find no more than 2 notes that sum to ${privateInput}.`);
        }
        const numInputNotes = notes.length;
        const totalNoteInputValue = notes.reduce((sum, note) => sum + note.value, BigInt(0));
        const inputNoteIndices = notes.map(n => n.index);
        const inputNotes = notes.map(n => new note_algorithms_1.TreeNote(n.owner.publicKey, n.value, n.assetId, n.owner.nonce, n.secret));
        const maxNoteIndex = Math.max(...inputNoteIndices, 0);
        // Add gibberish notes to ensure we have two notes.
        for (let i = notes.length; i < 2; ++i) {
            inputNoteIndices.push(maxNoteIndex + i); // notes can't have the same index
            inputNotes.push(note_algorithms_1.TreeNote.createFromEphPriv(publicKey, BigInt(0), assetId, nonce, this.createEphemeralPrivKey(), this.grumpkin));
        }
        const inputNotePaths = await Promise.all(inputNoteIndices.map(async (idx) => this.worldState.getHashPath(idx)));
        const changeValue = totalNoteInputValue > privateInput ? totalNoteInputValue - privateInput : BigInt(0);
        const outputNotes = [
            this.createNote(assetId, recipientPrivateOutput, newNoteOwner || id),
            this.createNote(assetId, changeValue + senderPrivateOutput, id),
        ];
        const claimNote = isDefiBridge
            ? this.createClaimNote(bridgeId, defiDepositValue, id)
            : {
                note: note_algorithms_1.ClaimNoteTxData.EMPTY,
                viewingKey: viewing_key_1.ViewingKey.EMPTY,
            };
        const dataRoot = this.worldState.getRoot();
        const inputOwner = inputOwnerAddress || address_1.EthAddress.ZERO;
        const outputOwner = outputOwnerAddress || address_1.EthAddress.ZERO;
        // For now, we will use the account key as the signing key (no account note required).
        const { privateKey } = userState.getUser();
        const message = client_proofs_1.computeSigningData([...inputNotes, ...outputNotes.map(n => n.note)], claimNote.note, inputNoteIndices[0], inputNoteIndices[1], inputOwner, outputOwner, publicInput, publicOutput, assetId, numInputNotes, id, privateKey, this.pedersen, this.noteAlgos);
        const signature = await signer.signMessage(message);
        const tx = new client_proofs_1.JoinSplitTx(publicInput, publicOutput, assetId, numInputNotes, inputNoteIndices, dataRoot, inputNotePaths, inputNotes, outputNotes.map(n => n.note), claimNote.note, privateKey, accountAliasId, accountIndex, accountPath, signingPubKey, signature, inputOwner, outputOwner);
        const viewingKeys = [isDefiBridge ? claimNote.viewingKey : outputNotes[0].viewingKey, outputNotes[1].viewingKey];
        return { tx, viewingKeys };
    }
    createNote(assetId, value, owner) {
        const ephKey = this.createEphemeralPrivKey();
        const note = note_algorithms_1.TreeNote.createFromEphPriv(owner.publicKey, value, assetId, owner.nonce, ephKey, this.grumpkin);
        const viewingKey = note.getViewingKey(ephKey, this.grumpkin);
        return { note, viewingKey };
    }
    createClaimNote(bridgeId, value, owner) {
        const ephKey = this.createEphemeralPrivKey();
        const note = note_algorithms_1.ClaimNoteTxData.createFromEphPriv(value, bridgeId, owner, ephKey, this.grumpkin);
        const viewingKey = note.getViewingKey(owner.publicKey, ephKey, this.grumpkin);
        return { note, viewingKey };
    }
    createEphemeralPrivKey() {
        return this.grumpkin.getRandomFr();
    }
}
exports.JoinSplitTxFactory = JoinSplitTxFactory;
