// @aztec/blockchain/price_feed/token_price_feed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenPriceFeed = void 0;
const ethers_1 = require("ethers");
const abi = [
    'function latestAnswer() public view returns(int256)',
    'function latestRound() public pure returns(uint256)',
    'function getRoundData(uint80 _roundId) public view returns(uint80,int256,uint256,uint256,uint80)',
];
class TokenPriceFeed {
    constructor(priceFeedContractAddress, provider) {
        this.contract = new ethers_1.Contract(priceFeedContractAddress.toString(), abi, provider);
    }
    async price() {
        return BigInt(await this.contract.latestAnswer());
    }
    async latestRound() {
        return BigInt(await this.contract.latestRound());
    }
    async getRoundData(roundId) {
        const [, answer, , updatedAt] = (await this.contract.getRoundData(roundId)) || [];
        return {
            roundId,
            price: BigInt(answer || 0),
            timestamp: +(updatedAt || 0),
        };
    }
    async getHistoricalPrice(roundId) {
        try {
            const data = await this.contract.getRoundData(roundId);
            return data ? BigInt(data[1]) : BigInt(0);
        }
        catch (e) {
            return BigInt(0);
        }
    }
}
exports.TokenPriceFeed = TokenPriceFeed;
