// @aztec/barretenberg/account_id/account_id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountId = void 0;
const address_1 = require("../address");
class AccountId {
    constructor(publicKey, nonce) {
        this.publicKey = publicKey;
        this.nonce = nonce;
    }
    static fromBuffer(id) {
        if (id.length !== 68) {
            throw new Error('Invalid id buffer.');
        }
        const publicKey = new address_1.GrumpkinAddress(id.slice(0, 64));
        const nonce = id.readUInt32BE(64);
        return new AccountId(publicKey, nonce);
    }
    static fromString(idStr) {
        const [match, publicKeyStr, nonceStr] = idStr.match(/^0x([0-9a-f]{128}) \(([0-9]+)\)$/i) || [];
        if (!match) {
            throw new Error('Invalid id string.');
        }
        const publicKey = address_1.GrumpkinAddress.fromString(publicKeyStr);
        return new AccountId(publicKey, +nonceStr);
    }
    static random() {
        const randomNonce = Math.floor(Math.random() * 2 ** 32);
        return new AccountId(address_1.GrumpkinAddress.randomAddress(), randomNonce);
    }
    equals(rhs) {
        return this.toBuffer().equals(rhs.toBuffer());
    }
    toBuffer() {
        const nonceBuf = Buffer.alloc(4);
        nonceBuf.writeUInt32BE(this.nonce);
        return Buffer.concat([this.publicKey.toBuffer(), nonceBuf]);
    }
    toString() {
        return `${this.publicKey.toString()} (${this.nonce})`;
    }
}
exports.AccountId = AccountId;
